{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#packages for utilzing the mutagenesis code\n",
    "import re #regular expressions\n",
    "from Bio import Entrez, SeqIO\n",
    "Entrez.email = 'sethfrazer.edu'\n",
    "from skbio import TabularMSA\n",
    "from skbio import Protein\n",
    "from skbio.alignment import global_pairwise_align_protein\n",
    "from Bio.Align import substitution_matrices\n",
    "import itertools\n",
    "#packages for utilizing deepBreaks\n",
    "import os\n",
    "import subprocess\n",
    "from deepBreaks.utils import load_obj\n",
    "from deepBreaks.preprocessing import read_data\n",
    "import joblib\n",
    "import pandas as pd\n",
    "import itertools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getAcc(accession, check):\n",
    "    if accession == \"manual\":\n",
    "\n",
    "        if check == 0:\n",
    "            manual = input(\"Enter Reference Sequence: \")\n",
    "        else:\n",
    "            manual = input(\"Enter Target Sequence: \")\n",
    "\n",
    "        return(manual)\n",
    "    \n",
    "    if accession == \"AncAgnatha\":\n",
    "        return(\"MNGTEGINFYVPLSNKTGLVRSPFEYPQYYLAEPWKYSALAAYMFFLILIGFPINFLTLFVTFKHKKLRQPLNYILVNLAVADLFMVCFGFTVTFYTAMNGYFVFGPTGCNIEGFFATLGGEVALWSLVVLAIERYIVVCKPMGNFRFASTHAIMGVAFTWIMALACAAPPLFGWSRYIPEGMQCSCGPDYYTLNPKYHNESYVIYMFVVHFLLPVTIIFFTYGRLICTVKEAAAQQQESASTQKAEKEVTRMVIIMVIGFLVCWVPYASVAFYIFTNKGADFSATLMTVPAFFSKSSALYNPIIYVLLNKQFRNCMITTICCGKNPFGDEDVSSSVSQSKTEVSSVSSSQVSPA\")\n",
    "    if accession == \"AncActino\":\n",
    "        return(\"MNGTEGINFYVPLSNKTGLVRSPFEYPQYYLAEPWKYSVLAAYMFFLILIGFPINFLTLLVTFKHKKLRQPLNYILVNLAVADLFMVCFGFTVTFYTAMNGYFVFGPTGCAIEGFFATLGGQVALWSLVVLAIERYIVVCKPMGNFRFASTHAIMGIAFTWIMALACAAPPLFGWSRYIPEGMQCSCGPDYYTLNPDYHNESYVMYMFVVHFLLPVTIIFFTYGRLICTVKEAAAQQQESASTQKAEKEVTRMVILMVIGFLVAWVPYASVAFWIFTNKGADFTATLMTVPAFFSKSSALYNPIIYVLLNKQFRNCMITTICCGKNPFGDEDVSSSVSQSKTEVSSVSSSQVSPA\")\n",
    "    if accession == \"AncTeleost\":\n",
    "        return(\"MNGTEGINFYVPLSNKTGLVRSPFEYPQYYLAEPWQYKLLAVYMFFLICFGFPINGLTLVVTAQHKKLRQPLNFILVNLAVAGLIMVCFGFTVTFYTAINGYFVFGPTGCAIEGFMATLGGQVALWSLVVLAIERYIVVCKPMGSFKFTSTHAIVGIAFTWIMALSCAAPPLFGWSRYIPEGMQCSCGPDYYTLNPDYHNESYVMYMFVVHFFLPVTIIFFTYGSLVCTVKAAAAQQQESASTQKAEKEVTRMVILMVLGFLVAWTPYASVAAWIFFNKGAAFTAQSMAVPAFFSKSSALYNPIIYVLLNKQFRNCMITTICCGKNPFGDEDVSSSVSQSKTEVSSVSSSQVSPA\")\n",
    "    if accession == \"AncClupeoc\":\n",
    "        return(\"MNGTEGINFYVPLSNKTGLVRSPFEYPQYYLAEPWQFKLLAVYMFFLICFGFPINGLTLVVTAQHKKLRQPLNFILVNLAVAGTIMVCFGFTVTFYTAINGYFVLGPTGCAIEGFMATLGGQVALWSLVVLAIERYIVVCKPMGSFKFSSSHAIAGIAFTWVMAMSCAAPPLFGWSRYIPEGMQCSCGPDYYTLNPEYNNESYVLYMFICHFCLPVTIIFFTYGSLVCTVKAAAAQQQESESTQKAEREVTRMVILMVLGFLVAWTPYASFAAWIFFNKGAAFSAQSMAVPAFFSKSSALYNPIIYVLLNKQFRNCMITTICCGKNPFGDEDVSSSVSQSKTEVSSVSSSQVSPA\")\n",
    "    if accession == \"AncCyprin1\":\n",
    "        return(\"MNGTEGINFYVPLSNKTGLVRSPFEYPQYYLAEPWQFKLLAVYMFFLICFGFPINGLTLVVTAQHKKLRQPLNFILVNLAVAGTIMVCFGFTVTFYTAINGYFVLGPTGCAIEGFMATLGGQVALWSLVVLAIERYIVVCKPMGSFKFSSSHAIAGIAFTWVMAMSCAAPPLFGWSRYIPEGMQCSCGPDYYTLNPEYNNESYVLYMFICHFCLPVTIIFFTYGSLVCTVKAAAAQQQESESTQKAEREVTRMVILMVLGFLVAWTPYASFAAWIFFNKGAAFSAQSMAVPAFFSKSSALYNPIIYVLLNKQFRNCMITTICCGKNPFGDEDVSSSVSQSKTEVSSVSSSQVSPA\")\n",
    "    if accession == \"AncCyprin2\":\n",
    "        return(\"MNGTEGINFYVPLSNKTGLVRSPFEYPQYYLAEPWQFKLLAVYMFFLICLGFPINGLTLVVTAQHKKLRQPLNFILVNLAVAGTIMVCFGFTVTFYTAINGYFVLGPTGCAIEGFMATLGGEVALWSLVVLAIERYIVVCKPMGSFKFSSSHAFAGIAFTWVMAMACAAPPLVGWSRYIPEGMQCSCGPDYYTLNPEYNNESYVLYMFICHFILPVTIIFFTYGRLVCTVKAAAAQQQESESTQKAEREVTRMVILMVLGFLVAWTPYATVAAWIFFNKGAAFSAQFMAVPAFFSKSSALYNPVIYVLLNKQFRNCMITTICCGKNPFGDEDVSSSVSQSKTEVSSVSSSQVSPA\")\n",
    "    if accession == \"AncEutelo1\":\n",
    "        return(\"MNGTEGINFYVPLSNKTGLVRSPFEYPQYYLAEPWQFKLLAVYMFFLICFGFPINGLTLVVTAQHKKLRQPLNFILVNLAVAGLIMVCFGFTVTFYTALNGYFVFGPLGCAIEGFMATLGGQVALWSLVVLAIERYIVVCKPMGSFKFTATHAAAGVAFTWIMAMSCAAPPLFGWSRYIPEGMQCSCGPDYYTLNPGYNNESYVMYMFTCHFCLPVTIIFFTYGSLVCTVKAAAAQQQESASTQKAEREVTRMCILMVLGFLVAWTPYASFAAWIFFNKGAAFTAQSMAVPAFFSKSSALFNPIIYVLMNKQFRNCMITTICCGKNPFGDEDVSSSVSQSKTEVSSVSSSQVSPA\")\n",
    "    if accession == \"AncEutelo2\":\n",
    "        return(\"MNGTEGINFYVPLSNKTGLVRSPFEYPQYYLADPWQFKLLAFYMFFLICTGFPINALTLVVTAQNKKLRQPLNFILVNLAVAGLIMVCFGFTVTIYSALNGYFVLGPLGCAIEGFMATLGGQVALWSLVVLAIERYIVVCKPMGSFKFTATHAAAGVAFTWIMAMSCAAPPLFGWSRYIPEGMQCSCGPDYYTLAPGYNNESYVMYMFTCHFCVPVFIIFFTYGSLVMTVKAAAAQQQDSASTQKAEKEVTRMCILMVLGFLVAWTPYASFAAWIFFNKGAAFTAMSMAIPAFFSKSSALFNPIIYVLMNKQFRNCMITTICCGKNPFGDEDVSSSVSQSKTEVSSVSSSQVSPA\")\n",
    "    if accession == \"AncEutelo3\":\n",
    "        return(\"MNGTEGINFYVPLSNKTGLVRSPFEYPQYYLADPWQFKLLAFYMFFLILTGFPINALTLVVTAQNKKLRQPLNFILVNLAVAGLIMVCFGFTVTIYSCLNGYFSLGPLGCAIEGFMATLGGQVSLWSLVVLAIERYIVVCKPMGSFKFTATHAAAGCAFTWIMAMSCAAPPLLGWSRYIPEGIQVSCGPDYYTLAPGYNNESYVMYMFTCHFCVPVFTIFFTYGSLVMTVKAAAAQQQDSASTQKAEKEVTRMCILMVLGFLLAWTPYASFAAWIFFNKGAAFTAMSMAIPSFFSKSSALFNPIIYILMNKQFRNCMITTICCGKNPFGDEDVSSSVSQSKTEVSSVSSSQVSPA\")\n",
    "    if accession == \"AncEutelo4\":\n",
    "        return(\"MNGTEGINFYVPLSNKTGLVRSPFEYPQYYMVDPIIYKILAFYMFFLICTGTPINGLTLFVTAQNKKLRQPLNYILVNLAVAGLIMCCFGFTITITSAVNGYFVLGPTFCAIEGFMATLGGEVALWSLVVLAVERYIVVCKPMGSFKFTGTHAAAGVLFTWIMAMACAAPPLFGWSRYIPEGMQCSCGPDYYTLAPGFNNESYVIYMFVVHFFVPVFIIFFTYGSLVMTVKAAAAQQQDSASTQKAEKEVTRMCVLMVMGFLVAWTPYASFAGWIFLNKGAAFTALTAAIPAFFAKSSALYNPVIYVLMNKQFRNCMITTICCGKNPFGDEDVSSSVSQSKTEVSSVSSSQVSPA\")\n",
    "    if accession == \"AncTetrapod\":\n",
    "        return(\"MNGTEGINFYVPLSNKTGLVRSPFEYPQYYLAEPWKYSVLCAYMFFLILIGFPINFLTLLVTFKHKKLRQPLNYILVNLAVADLFMVCFGFTVTFYTAMNGYFVFGPTGCAIEGFFATLGGQVALWSLVVLAIERYIVVCKPMGNFRFASTHAIMGIAFTWIMALACAAPPLFGWSRYIPEGMQCSCGPDYYTLNPDYHNESYVMYMFVVHFLLPVTIIFFTYGRLICKVKEAAAQQQESASTQKAEKEVTRMVILMVIGFLVAWVPYASVAFWIFTNKGADFTATLMTVPAFFSKSSALYNPIIYVLLNKQFRNCMITTICCGKNPFGDEDVSSSVSQSKTEVSSVSSSQVSPA\")\n",
    "    if accession == \"AncSaurops\":\n",
    "        return(\"MNGTEGINFYVPLSNKTGLVRSPFEYPQYYLAEPWKYKVVCCYIFFLISTGLPINFLTLLVTFKHKKLRQPLNYILVNLAVADLFMACFGFTVTFYTAWNGYFVFGPIGCAIEGFFATLGGQVALWSLVVLAIERYIVVCKPMGNFRFSATHALMGIAFTWVMAFSCAAPPLFGWSRYIPEGMQCSCGPDYYTLNPDYHNESYVLYMFVVHFIIPVVVIFFSYGRLICKVREAAAQQQESASTQKAEKEVTRMVILMVLGFLLAWTPYAVVAFWIFTNKGADFTATLMTVPAFFSKSSSLYNPIIYVLMNKQFRNCMITTICCGKNPFGDEDVSSSVSQSKTEVSSVSSSQVSPA\")\n",
    "    if accession == \"AncSquamat\":\n",
    "        return(\"MNGTEGINFYVPLSNKTGLVRSPFEYPQYYLAEPWKYKVVCCYIFFLISTGLPINFLTLLVTFKHKKLRQPLNYILVNLAVADLFMACFGFTVTFYTAWNGYFVFGPIGCAIEGFFATLGGQVALWSLVVLAIERYIVVCKPMGNFRFSATHALMGIAFTWFMAFSCAAPPLFGWSRYIPEGMQCSCGPDYYTLNPDYHNESYVLYMFVVHFIIPVVVIFFSYGRLICKVREAAAQQQESASTQKAEKEVTRMVILMVLGFLLAWTPYAVVAFWIFTNKGADFSATLMTVPAFFSKSSSLYNPIIYVLMNKQFRNCMITTICCGKNPFGDEDVSSSVSQSKTEVSSVSSSQVSPA\")\n",
    "    if accession == \"CNUCHROD\":\n",
    "        return(\"MNGTEGQDFYVPMSNKTGVVRSPFEYPQYYLAEPWKFSALAAYMFMLILLGFPINFLTLYVTIQHKKLRTPLNYILLNLAVANLFMVFGGFTTTMYTSMNGYFVFGVTGCYIEGFFATLGGEIALWSLVVLAIERYVVVCKPMSNFRFGENHAILGVAFSWIMALACAAPPLFGWSRYIPEGMQCSCGIDYYTLKPEVNNESFVIYMFVVHFMIPLLIIFFCYGNLVCTVKEAAAQQQESATTQKAEKEVTRMVIIMVIAFLICWVPYASVAFYIFTNQGSDFGPIFMTIPAFFAKSSAIYNPVIYIVMNKQFRNCMITTLCCGKNPLGDEDTSAGKTETSSVSTSQVSPA\")\n",
    "    if accession == \"ORCAROD\":\n",
    "        return(\"MNGTEGLNFYVPFSNKTGVVRSPFEYPQYYLAEPWQFSVLAAYMFLLIVLGFPINFLTLYVTVQHKKLRTPLNYILLNLAVANLFMVFGGFTTTLYTSLHAYFIFGPTGCNLEGFFATLGGEIALWSLVVLAIERYVVVCKPMSNFRFGENHAIMGLALTWIMAMACAAPPLVGWSRYIPEGMQCSCGIDYYTLSPEVNNESFVIYMFVVHFTIPLVIIFFCYGQLVFTVKEAAAQQQESATTQKAEKEVTRMVIIMVVAFLICWVPYASVAFYIFTHQGSDFGPIFMTIPSFFAKSSSIYNPVIYIMMNKQFRNCMLTTLCCGRNPLGDDEASTTASKTETSQVAPA\")\n",
    "    if accession == \"HIPPOROD\":\n",
    "        return(\"MNGTEGPNFYVPFSNKTGVVRSPFEYPQYYLAEPWQFSMLAAYMFLLIVLGFPINFLTLYVTVQHKKLRTPLNYILLNLAVADLFMVFGGFTTTLYTSMHGYFIFGPTGCNLEGFFATLGGEIALWSLVVLAIERYVVVCKPMSNFRFGENHAIMGLALTWVMALACAAPPLVGWSRYIPEGMQCSCGIDYYTLKPEVNNESFVIYMFVVHFSIPMVIIFFCYGQLVFTVKEAAAQQQESATTQKAEKEVTRMVIIMVIAFLICWLPYAGVAFYIFTHQGSDFGPIFMTIPAFFAKTSSVYNPVIYIMMNKQFRNCMLTTLCCGKNPLGDDEASTTASKTETSQVAPA\")\n",
    "    try:\n",
    "        handle = Entrez.efetch(db=\"nucleotide\", id=accession, rettype=\"gb\", retmode=\"text\")\n",
    "        record = SeqIO.read(handle, \"gb\")\n",
    "        handle.close()\n",
    "        for i,feature in enumerate(record.features):\n",
    "            if feature.type=='CDS':\n",
    "                aa = feature.qualifiers['translation'][0]\n",
    "    except:\n",
    "        manual = input(f\"Accession# {accession} - Not Recognized by NCBI\\nPlease Enter Target Sequence to Continue: \")\n",
    "        return(manual)\n",
    "    return(aa)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getCombos(mut_list, wt_seq, mut_acc_file):\n",
    "    wt_seq = f'{wt_seq}_'\n",
    "    x = 1\n",
    "    mut_combos = []\n",
    "    while x in range(len(mut_list)+1):\n",
    "        combinations = list(itertools.combinations(mut_list, x))\n",
    "\n",
    "        for combo in combinations:\n",
    "            acc = wt_seq\n",
    "\n",
    "            check = 1\n",
    "            for i in combo:\n",
    "                if check == x:    \n",
    "                    acc+=i\n",
    "                else:\n",
    "                    acc = f'{acc}{i},'\n",
    "                check+=1\n",
    "            print(acc)\n",
    "\n",
    "            mut_combos.append(acc)\n",
    "        x +=1\n",
    "    #print(mut_combos)\n",
    "    #print(len(mut_combos))\n",
    "\n",
    "    with open(mut_acc_file, 'w') as f:\n",
    "        for mut_acc in mut_combos:\n",
    "            f.write(f'{mut_acc}\\n')\n",
    "    \n",
    "    return(mut_combos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getMutAcc_fromCSV(mut_file, acc_name):\n",
    "    # Replace 'your_file.csv' with the path to your CSV file    \n",
    "    df = pd.read_csv(mut_file)\n",
    "\n",
    "    #  Access and print the column headers\n",
    "    column_headers = list(df.columns)\n",
    "    print(column_headers)\n",
    "\n",
    "    base_acc = acc_name\n",
    "    list_of_accs = []\n",
    "\n",
    "    for row in df.itertuples():\n",
    "        current_acc = base_acc\n",
    "        check = 0\n",
    "        # Iterate through each column value\n",
    "        for i in range(1, len(row)):\n",
    "            if row[i] == 1:\n",
    "                if check == 0:\n",
    "                    current_acc += '_' + column_headers[i-1]\n",
    "                    check+=1\n",
    "                else:\n",
    "                    current_acc += ',' + column_headers[i-1]\n",
    "        \n",
    "        # Append the constructed string to the list and reset current_acc\n",
    "        if current_acc != base_acc:\n",
    "            list_of_accs.append(current_acc)\n",
    "            \n",
    "        current_acc = base_acc\n",
    "    print(list_of_accs)\n",
    "    return(list_of_accs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getMut(mut_combos,output_file, out_format='fasta', allow_wt = True, raccession='NM_001014890.2'):\n",
    "    switch = 0\n",
    "    bovine = Protein(getAcc(raccession, switch))\n",
    "    switch+=1\n",
    "\n",
    "    try:\n",
    "        mut_combos = open(mut_combos).readlines()\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "\n",
    "    #Fetch sequences to manipulate and align\n",
    "    test = 0\n",
    "    if out_format != 'fasta':\n",
    "        with open(output_file, 'w') as f:\n",
    "            f.write(f'Accessions\\tMutant_Seqs\\n')\n",
    "    else:\n",
    "        with open(output_file, 'w') as f:\n",
    "            f.write('')\n",
    "    for lines in mut_combos:     \n",
    "    \n",
    "        if len(lines.split(\"_\")[0]) == 2:\n",
    "            try:\n",
    "                accession = f'{lines.split(\"_\")[0]}_{lines.split(\"_\")[1]}'\n",
    "                muts = lines.split('_')[2]\n",
    "            except:\n",
    "                accession = f'{lines.split(\"_\")[0]}_{lines.split(\"_\")[1]}'\n",
    "                muts = []\n",
    "\n",
    "        elif '_' not in lines:\n",
    "            accession = lines\n",
    "            muts = []\n",
    "            wt = Protein(getAcc(accession, switch))\n",
    "\n",
    "        else:\n",
    "            accession = f'{lines.split(\"_\")[0]}'\n",
    "            muts = lines.split('_')[1]\n",
    "\n",
    "        accession = accession.replace('\\n','')\n",
    "        wt = Protein(getAcc(accession, switch))\n",
    "\n",
    "        substitution_matrix = substitution_matrices.load(\"BLOSUM45\")\n",
    "\n",
    "        ##Simple example for testing\n",
    "        #wt=Protein(\"ABCDEF\")\n",
    "        #bovine=Protein(\"ABCDEF\")\n",
    "\n",
    "        alignment, score, start_end_positions = global_pairwise_align_protein(bovine, wt, substitution_matrix=substitution_matrix, gap_open_penalty=20)\n",
    "        dic = alignment.to_dict()\n",
    "        aligned_bovine = dic[0]\n",
    "        aligned_wt = dic[1]\n",
    "\n",
    "        #if test == 0:\n",
    "            #print(f'This is the aligned bovine:\\n{aligned_bovine}')\n",
    "            #print(f'This is {accession} aligned to bovine:\\n{aligned_wt}')\n",
    "        \n",
    "        #Check if any mutation is entered\n",
    "        full_mutation = muts\n",
    "        if len(muts) == 0:\n",
    "            if allow_wt == True:\n",
    "                print(\"You've requested a WT Sequence\")\n",
    "                if out_format == 'fasta':\n",
    "                    with open(output_file, 'a') as f:\n",
    "                        f.write(f'>{accession}\\n{wt}\\n')\n",
    "\n",
    "                else:\n",
    "                    with open(output_file, 'a') as f:\n",
    "                        f.write(f'{accession}\\t{wt}\\n')\n",
    "            else:\n",
    "                raise Exception(\"ERROR: Expecting mutation name in the format of XaY where X=old amino acid a=number Y=new amino acid, but received WT sequence\")\n",
    "        else: \n",
    "            #Check if single or multiple mutations entered\n",
    "            #print(f'This is the set of mutations: {muts}')\n",
    "\n",
    "            if \",\" in muts :\n",
    "                mutations = muts.split(\",\")\n",
    "            else :\n",
    "                #declare empty list and append single mutation so later can be used as list if only one mutation\n",
    "                mutations = []\n",
    "                mutations.append(muts)\n",
    "\n",
    "            mutated = aligned_wt #need to copy original sequence and accumulate mutations\n",
    "            for mutation in mutations:\n",
    "                #print(f'Checking mutation: {mutation}')\n",
    "                #check that mutation is formatted properly\n",
    "                MutationRegex = re.compile(r'([A-Z])(\\d+)([A-Z])')\n",
    "                if re.match(MutationRegex, mutation):\n",
    "                    mu = MutationRegex.search(mutation)\n",
    "                    #print(f'Here is the RegEx results: {mu}')\n",
    "                    old = mu.group(1)\n",
    "                    mutsite = int(mu.group(2))\n",
    "                    new = mu.group(3)\n",
    "                elif 'del' in mutation:\n",
    "                    MutationRegex = re.compile(r'([A-Z])(\\d+)([a-z]+)')\n",
    "                    mu = MutationRegex.search(mutation)\n",
    "                    old = mu.group(1)\n",
    "                    mutsite = int(mu.group(2))\n",
    "                    new = mu.group(3)\n",
    "                elif 'ins' in mutation:\n",
    "                    MutationRegex = re.compile(r'([a-z]+)(\\d+)([A-Z])')\n",
    "                    mu = MutationRegex.search(mutation)\n",
    "                    old = mu.group(1)\n",
    "                    mutsite = int(mu.group(2))\n",
    "                    new = mu.group(3)\n",
    "                else :\n",
    "                    raise Exception(\"ERROR: Expecting mutation name in the format of XaY where X=old amino acid a=number Y=new amino acid\")\n",
    "\n",
    "                #substract one to account for index starting at zero\n",
    "                #print(mutsite)\n",
    "                mutsite = mutsite-1\n",
    "                #print(mutsite)\n",
    "\n",
    "                #Count gaps to not count them when finding original site to change\n",
    "                gaps = aligned_bovine[:mutsite].count('-')\n",
    "                #print(gaps)\n",
    "\n",
    "\n",
    "                if new == 'del':\n",
    "                    new = ''\n",
    "                if old == 'ins':\n",
    "                    mutated = str(mutated[:mutsite]) + new + str(mutated[mutsite:])\n",
    "                else:\n",
    "                    #match = str(mutated[mutsite+gaps])\n",
    "                    match = str(mutated[mutsite])\n",
    "\n",
    "                    #print(match)\n",
    "                    if match != old :\n",
    "                        try: \n",
    "                            match = str(mutated[mutsite+gaps])\n",
    "                            if match != old: \n",
    "                                    raise Exception()\n",
    "                            else:\n",
    "                                mutated = str(mutated[:mutsite+gaps]) + new + str(mutated[mutsite+gaps+1:]) \n",
    "                        except:\n",
    "                            print(f'The AA at site {mutsite+1} is supposed to be {old} but appears to be {match}')\n",
    "                            raise Exception(\"ERROR: The mutation label does not match the sequence\")\n",
    "                    else: \n",
    "                        mutated = str(mutated[:mutsite]) + new + str(mutated[mutsite+1:]) \n",
    "\n",
    "            mutated = mutated.replace('-','')\n",
    "            mutated.replace('\\n','')\n",
    "            newacc = accession + \"_\" + full_mutation\n",
    "            newacc = newacc.replace('\\n','')\n",
    "\n",
    "            if out_format == 'fasta':\n",
    "                with open(output_file, 'a') as f:\n",
    "                    f.write(f'>{newacc}\\n{mutated}\\n')\n",
    "                    print(newacc, end='\\t\\t')\n",
    "                    #print(mutated)\n",
    "            else:\n",
    "                with open(output_file, 'a') as f:\n",
    "                    f.write(f'{newacc}\\t{mutated}\\n')\n",
    "                    print(newacc, end='\\t\\t')\n",
    "                    #print(mutated)\n",
    "\n",
    "    return(newacc , mutated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#mut_acc_list = getMutAcc_fromCSV('human_rho_mutatnts.csv', 'AncBoreotheria')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "raccession = 'NM_001014890.2'\n",
    "input_file = 'temp.txt'\n",
    "output_file = 'temp.tsv'\n",
    "getMut(mut_combos=input_file,output_file=output_file,out_format='tsv',allow_wt=True,raccession=raccession)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "wds_mut_list = ['A132S', 'N151S', 'G174D', 'P180R', 'F228L', 'F287L']\n",
    "wt = 'NM_001280659'\n",
    "mut_acc_file = 'dolphin_muts_csts.txt'\n",
    "raccession = 'NM_001014890'\n",
    "mutant_seq_file = 'dolphin_muts_csts.fasta'\n",
    "out_format = 'fasta'\n",
    "combos = getCombos(wds_mut_list, wt, mut_acc_file)\n",
    "getMut(combos,raccession,mutant_seq_file, out_format)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "raccession = 'NM_001014890.2'\n",
    "input_file = 'bovine_mutagensis_accs.txt'\n",
    "output_file = 'bovine_mutagenesis.fasta'\n",
    "getMut(mut_combos=input_file,output_file=output_file,out_format='fasta',allow_wt=True,raccession=raccession)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "deepBreaks",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.17"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
