{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color=#c51b8a>VPOD 'Mine-n-Match':</font>\n",
    "## <font color=#c994c7>Part 1 Objective</font> - Use Species Names from Microspectrophotemetry Data Sheet to Query NCBI for All Related Opsin Sequences  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import datetime\n",
    "import time\n",
    "import json\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from deepBreaks.preprocessing import read_data\n",
    "from Bio import Entrez, SeqIO\n",
    "email = 'sethfrazer@ucsb.edu'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from mnm_scripts.mine_n_match_functions import ncbi_fetch_opsins, merge_accessory_dbs, fasta_to_dataframe, correct_species_name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color=#c51b8a>Load data-tables with all of the species and Lambda Max data from accessory lmax databases</font> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color=#c994c7>VPOD Single Cell Microspectrophotmetry (SCP) Datatable </font>\n",
    "### In this case our dataframe does not have full species name in one column so we must create a list by directly combining the genus and species names. Then filter to create a list of all unique species names \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "report_dir = './data_sources/lmax'\n",
    "species_list = []\n",
    "df_list = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>phylum</th>\n",
       "      <th>CellType</th>\n",
       "      <th>CellSubType</th>\n",
       "      <th>LambdaMax</th>\n",
       "      <th>error</th>\n",
       "      <th>Chromophore</th>\n",
       "      <th>Stage</th>\n",
       "      <th>useable</th>\n",
       "      <th>expid</th>\n",
       "      <th>opsinid</th>\n",
       "      <th>refid</th>\n",
       "      <th>Notes</th>\n",
       "      <th>Column1</th>\n",
       "      <th>isGenusInHetero</th>\n",
       "      <th>isSpeciesInHetero</th>\n",
       "      <th>isInHetero</th>\n",
       "      <th>isVert</th>\n",
       "      <th>Full_Species</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>maxid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Acipenser</td>\n",
       "      <td>transmontanus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>rod</td>\n",
       "      <td>NaN</td>\n",
       "      <td>540.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>adult</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>463.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>partial cds but is rod opsin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Acipenser transmontanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Acipenser</td>\n",
       "      <td>transmontanus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>cone</td>\n",
       "      <td>NaN</td>\n",
       "      <td>464.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>adult</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Acipenser transmontanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Acipenser</td>\n",
       "      <td>transmontanus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>cone</td>\n",
       "      <td>NaN</td>\n",
       "      <td>531.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>adult</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Acipenser transmontanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Acipenser</td>\n",
       "      <td>transmontanus</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>cone</td>\n",
       "      <td>NaN</td>\n",
       "      <td>605.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>adult</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Acipenser transmontanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Acipenser</td>\n",
       "      <td>medirostris</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>rod</td>\n",
       "      <td>NaN</td>\n",
       "      <td>540.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>adult</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Acipenser medirostris</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Genus        Species    phylum CellType CellSubType  LambdaMax  \\\n",
       "maxid                                                                       \n",
       "1      Acipenser  transmontanus  Chordata      rod         NaN      540.0   \n",
       "2      Acipenser  transmontanus  Chordata     cone         NaN      464.0   \n",
       "3      Acipenser  transmontanus  Chordata     cone         NaN      531.0   \n",
       "4      Acipenser  transmontanus  Chordata     cone         NaN      605.0   \n",
       "5      Acipenser    medirostris  Chordata      rod         NaN      540.0   \n",
       "\n",
       "       error Chromophore  Stage  useable  expid  opsinid  refid  \\\n",
       "maxid                                                             \n",
       "1        NaN         NaN  adult      1.0    NaN    463.0   30.0   \n",
       "2        NaN         NaN  adult      0.0    NaN      0.0   30.0   \n",
       "3        NaN         NaN  adult      0.0    NaN      0.0   30.0   \n",
       "4        NaN         NaN  adult      0.0    NaN      0.0   30.0   \n",
       "5        NaN         NaN  adult      0.0    NaN      0.0   43.0   \n",
       "\n",
       "                              Notes Column1  isGenusInHetero  \\\n",
       "maxid                                                          \n",
       "1      partial cds but is rod opsin     NaN              NaN   \n",
       "2                               NaN     NaN              NaN   \n",
       "3                               NaN     NaN              NaN   \n",
       "4                               NaN     NaN              NaN   \n",
       "5                               NaN     NaN              NaN   \n",
       "\n",
       "       isSpeciesInHetero  isInHetero  isVert             Full_Species  \n",
       "maxid                                                                  \n",
       "1                    NaN           0     1.0  Acipenser transmontanus  \n",
       "2                    NaN           0     NaN  Acipenser transmontanus  \n",
       "3                    NaN           0     NaN  Acipenser transmontanus  \n",
       "4                    NaN           0     NaN  Acipenser transmontanus  \n",
       "5                    NaN           0     NaN    Acipenser medirostris  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scp_file = f'{report_dir}/vpod/scp_cleaned.txt'\n",
    "scp_df = pd.read_csv(scp_file, index_col=0, sep='\\t')\n",
    "scp_df['Full_Species'] = (scp_df['Genus'] + ' ' + scp_df['Species']).to_list()\n",
    "df_list.append(scp_df)\n",
    "scp_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_list+=scp_df['Full_Species'].to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color=#c994c7>Longcore - 'Animal Photopigments' Datatable</font>\n",
    "\n",
    "### In this next case our dataframe has the full species name in one column so we can create a list directly and filter to create a list of all unique species names\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Full_Species</th>\n",
       "      <th>Type</th>\n",
       "      <th>LambdaMax</th>\n",
       "      <th>Band</th>\n",
       "      <th>Oil</th>\n",
       "      <th>Nocturnal Activity</th>\n",
       "      <th>Diurnal Activity</th>\n",
       "      <th>Reference</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>longcore_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Arachnida</td>\n",
       "      <td>Aranae</td>\n",
       "      <td>Araneidae</td>\n",
       "      <td>Argiope amoena</td>\n",
       "      <td>NaN</td>\n",
       "      <td>360</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Yamashita &amp; Tateda 1978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Arachnida</td>\n",
       "      <td>Aranae</td>\n",
       "      <td>Araneidae</td>\n",
       "      <td>Argiope amoena</td>\n",
       "      <td>NaN</td>\n",
       "      <td>490</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Yamashita &amp; Tateda 1978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Arachnida</td>\n",
       "      <td>Aranae</td>\n",
       "      <td>Araneidae</td>\n",
       "      <td>Argiope amoena</td>\n",
       "      <td>NaN</td>\n",
       "      <td>540</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Yamashita &amp; Tateda 1978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Arachnida</td>\n",
       "      <td>Aranae</td>\n",
       "      <td>Araneidae</td>\n",
       "      <td>Argiope bruennichi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>360</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Yamashita &amp; Tateda 1978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Arachnida</td>\n",
       "      <td>Aranae</td>\n",
       "      <td>Araneidae</td>\n",
       "      <td>Argiope bruennichi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>490</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Yamashita &amp; Tateda 1978</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Phylum      Class   Order     Family        Full_Species  \\\n",
       "longcore_id                                                                 \n",
       "0            Arthropoda  Arachnida  Aranae  Araneidae      Argiope amoena   \n",
       "1            Arthropoda  Arachnida  Aranae  Araneidae      Argiope amoena   \n",
       "2            Arthropoda  Arachnida  Aranae  Araneidae      Argiope amoena   \n",
       "3            Arthropoda  Arachnida  Aranae  Araneidae  Argiope bruennichi   \n",
       "4            Arthropoda  Arachnida  Aranae  Araneidae  Argiope bruennichi   \n",
       "\n",
       "            Type  LambdaMax Band  Oil Nocturnal Activity Diurnal Activity  \\\n",
       "longcore_id                                                                 \n",
       "0            NaN        360  NaN  NaN                  Y                Y   \n",
       "1            NaN        490  NaN  NaN                  Y                Y   \n",
       "2            NaN        540  NaN  NaN                  Y                Y   \n",
       "3            NaN        360  NaN  NaN                  Y                Y   \n",
       "4            NaN        490  NaN  NaN                  Y                Y   \n",
       "\n",
       "                           Reference  \n",
       "longcore_id                           \n",
       "0            Yamashita & Tateda 1978  \n",
       "1            Yamashita & Tateda 1978  \n",
       "2            Yamashita & Tateda 1978  \n",
       "3            Yamashita & Tateda 1978  \n",
       "4            Yamashita & Tateda 1978  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "longcore_file = f'{report_dir}/longcore_data/AnimalPhotopigmentsV1_1.csv'\n",
    "longcore_df = pd.read_csv(longcore_file, index_col=0)\n",
    "df_list.append(longcore_df)\n",
    "longcore_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_list+=longcore_df['Full_Species'].to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color=#c994c7>Murphy and Westerman Datatable</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Full_Species</th>\n",
       "      <th>LambdaMax</th>\n",
       "      <th>SD</th>\n",
       "      <th>N</th>\n",
       "      <th>refid</th>\n",
       "      <th>Lineage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>murwes_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aglais urticae</td>\n",
       "      <td>360.0</td>\n",
       "      <td>–</td>\n",
       "      <td>†</td>\n",
       "      <td>[1]</td>\n",
       "      <td>Arthropod</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Agraulis vanillae</td>\n",
       "      <td>370.0</td>\n",
       "      <td>–</td>\n",
       "      <td>†</td>\n",
       "      <td>[2]</td>\n",
       "      <td>Arthropod</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Alima pacifica</td>\n",
       "      <td>467.0</td>\n",
       "      <td>–</td>\n",
       "      <td>5</td>\n",
       "      <td>[4]</td>\n",
       "      <td>Arthropod</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Antheraea polyphemus</td>\n",
       "      <td>355.0</td>\n",
       "      <td>–</td>\n",
       "      <td>20</td>\n",
       "      <td>[5]</td>\n",
       "      <td>Arthropod</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Apodemia mormo</td>\n",
       "      <td>556.0</td>\n",
       "      <td>–</td>\n",
       "      <td>†</td>\n",
       "      <td>[3]</td>\n",
       "      <td>Arthropod</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Full_Species  LambdaMax SD   N refid    Lineage\n",
       "murwes_id                                                         \n",
       "0                Aglais urticae      360.0  –   †   [1]  Arthropod\n",
       "1             Agraulis vanillae      370.0  –   †   [2]  Arthropod\n",
       "2                Alima pacifica      467.0  –   5   [4]  Arthropod\n",
       "3          Antheraea polyphemus      355.0  –  20   [5]  Arthropod\n",
       "4                Apodemia mormo      556.0  –   †   [3]  Arthropod"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "murphy_westerman_file = f'{report_dir}/murphy_westerman/Murphy and Westerman.csv'\n",
    "murphy_westerman_df = pd.read_csv(murphy_westerman_file, index_col=0)\n",
    "df_list.append(murphy_westerman_df)\n",
    "murphy_westerman_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_list+=murphy_westerman_df['Full_Species'].to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color=#c994c7>Caves 'Fish' Datatable</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>LambdaMax</th>\n",
       "      <th>refid</th>\n",
       "      <th>Full_Species</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>caves_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acipenseridae</td>\n",
       "      <td>Acipenser</td>\n",
       "      <td>baerii</td>\n",
       "      <td>549</td>\n",
       "      <td>[22]</td>\n",
       "      <td>Acipenser baerii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Acipenseridae</td>\n",
       "      <td>Acipenser</td>\n",
       "      <td>medirostrus</td>\n",
       "      <td>540</td>\n",
       "      <td>[60]</td>\n",
       "      <td>Acipenser medirostrus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Acipenseridae</td>\n",
       "      <td>Acipenser</td>\n",
       "      <td>ruthenus</td>\n",
       "      <td>545</td>\n",
       "      <td>[21]</td>\n",
       "      <td>Acipenser ruthenus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Acipenseridae</td>\n",
       "      <td>Acipenser</td>\n",
       "      <td>transmontanus</td>\n",
       "      <td>540</td>\n",
       "      <td>[39] ; [58]</td>\n",
       "      <td>Acipenser transmontanus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Acipenseridae</td>\n",
       "      <td>Scaphirhynchus</td>\n",
       "      <td>albus</td>\n",
       "      <td>538</td>\n",
       "      <td>[60]</td>\n",
       "      <td>Scaphirhynchus albus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Family           Genus        Species  LambdaMax  \\\n",
       "caves_id                                                            \n",
       "0         Acipenseridae       Acipenser         baerii        549   \n",
       "1         Acipenseridae       Acipenser    medirostrus        540   \n",
       "2         Acipenseridae       Acipenser       ruthenus        545   \n",
       "3         Acipenseridae       Acipenser  transmontanus        540   \n",
       "4         Acipenseridae  Scaphirhynchus          albus        538   \n",
       "\n",
       "                refid             Full_Species  \n",
       "caves_id                                        \n",
       "0                [22]         Acipenser baerii  \n",
       "1                [60]    Acipenser medirostrus  \n",
       "2                [21]       Acipenser ruthenus  \n",
       "3         [39] ; [58]  Acipenser transmontanus  \n",
       "4                [60]     Scaphirhynchus albus  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "caves_fish_db_file = f'{report_dir}/caves/caves_fish_db.csv'\n",
    "caves_df = pd.read_csv(caves_fish_db_file, index_col=0)\n",
    "caves_df['Full_Species'] = (caves_df['Genus'] + ' ' + caves_df['Species']).to_list()\n",
    "df_list.append(caves_df)\n",
    "caves_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_list+=caves_df['Full_Species'].to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color=#c994c7>Morgan Porter's Datatables</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Full_Species</th>\n",
       "      <th>Habitat</th>\n",
       "      <th>method</th>\n",
       "      <th>LambdaMax</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Accession</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>porter2005_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Artemia salina</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EON</td>\n",
       "      <td>410</td>\n",
       "      <td>(Hertel 1972)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Daphnia magna</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VC</td>\n",
       "      <td>348, 434, 525, 608</td>\n",
       "      <td>(Smith and Macagno 1990)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Acartia tonsa</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BP</td>\n",
       "      <td>450-520</td>\n",
       "      <td>(Stearns and Forward 1984)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Balanus amphitrite</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERG</td>\n",
       "      <td>532</td>\n",
       "      <td>(Hillman et al. 1973)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Balanus amphitrite</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MSP</td>\n",
       "      <td>532</td>\n",
       "      <td>(Minke and Kirschfield 1978)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Full_Species Habitat method           LambdaMax  \\\n",
       "porter2005_id                                                          \n",
       "0                  Artemia salina     NaN    EON                 410   \n",
       "1                   Daphnia magna     NaN     VC  348, 434, 525, 608   \n",
       "2                   Acartia tonsa     NaN     BP             450-520   \n",
       "3              Balanus amphitrite     NaN    ERG                 532   \n",
       "4              Balanus amphitrite     NaN    MSP                 532   \n",
       "\n",
       "                                  Reference Accession  \n",
       "porter2005_id                                          \n",
       "0                             (Hertel 1972)       NaN  \n",
       "1                  (Smith and Macagno 1990)       NaN  \n",
       "2                (Stearns and Forward 1984)       NaN  \n",
       "3                     (Hillman et al. 1973)       NaN  \n",
       "4              (Minke and Kirschfield 1978)       NaN  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "porter_file1 = f'{report_dir}/megan_porter/megan_porter_extract_clean_2005.csv'\n",
    "porter_df1 = pd.read_csv(porter_file1, index_col=0)\n",
    "porter_df2 = pd.read_csv(f'{report_dir}/megan_porter/porter_2006_table1_clean.tsv', sep='\\t', index_col=0)\n",
    "df_list.append(porter_df1)\n",
    "df_list.append(porter_df2)\n",
    "porter_df1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Full_Species</th>\n",
       "      <th>Accession</th>\n",
       "      <th>LambdaMax</th>\n",
       "      <th>Reference</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>porter2006_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Loligo forbesii</td>\n",
       "      <td>X56788</td>\n",
       "      <td>494</td>\n",
       "      <td>Morris et al. (1993)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Loligo pealii</td>\n",
       "      <td>AY450853</td>\n",
       "      <td>493</td>\n",
       "      <td>Brown PK and Brown PS (1958), Hubbard and St G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Loligo subulata</td>\n",
       "      <td>Z49108</td>\n",
       "      <td>499</td>\n",
       "      <td>Morris et al. (1993)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sepia officinalis</td>\n",
       "      <td>AF000947</td>\n",
       "      <td>492</td>\n",
       "      <td>Brown PK and Brown PS (1958)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Todarodes pacificus</td>\n",
       "      <td>X70498</td>\n",
       "      <td>480</td>\n",
       "      <td>Naito et al. (1981)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Full_Species Accession LambdaMax  \\\n",
       "porter2006_id                                            \n",
       "0                  Loligo forbesii    X56788       494   \n",
       "1                    Loligo pealii  AY450853       493   \n",
       "2                  Loligo subulata    Z49108       499   \n",
       "3                Sepia officinalis  AF000947       492   \n",
       "4              Todarodes pacificus    X70498       480   \n",
       "\n",
       "                                                       Reference  \n",
       "porter2006_id                                                     \n",
       "0                                           Morris et al. (1993)  \n",
       "1              Brown PK and Brown PS (1958), Hubbard and St G...  \n",
       "2                                           Morris et al. (1993)  \n",
       "3                                   Brown PK and Brown PS (1958)  \n",
       "4                                            Naito et al. (1981)  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "porter_df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_list += (porter_df1['Full_Species'].to_list() + porter_df2['Full_Species'].to_list())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color=#c994c7>J. Van Der Kooi's Datatable</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>First</th>\n",
       "      <th>Seco</th>\n",
       "      <th>Third</th>\n",
       "      <th>Fourth</th>\n",
       "      <th>Fifth</th>\n",
       "      <th>Sixth</th>\n",
       "      <th>Remaining</th>\n",
       "      <th>Full_Species</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>kooi_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Pieridae</td>\n",
       "      <td>Colias</td>\n",
       "      <td>erate</td>\n",
       "      <td>360</td>\n",
       "      <td>440</td>\n",
       "      <td>480</td>\n",
       "      <td>580.0</td>\n",
       "      <td>660.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Colias erate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Pieridae</td>\n",
       "      <td>Colias</td>\n",
       "      <td>erate</td>\n",
       "      <td>360</td>\n",
       "      <td>430</td>\n",
       "      <td>460</td>\n",
       "      <td>580.0</td>\n",
       "      <td>640.0</td>\n",
       "      <td>620.0</td>\n",
       "      <td>660</td>\n",
       "      <td>Colias erate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Pieridae</td>\n",
       "      <td>Leptidae</td>\n",
       "      <td>amurensis</td>\n",
       "      <td>360</td>\n",
       "      <td>440</td>\n",
       "      <td>450</td>\n",
       "      <td>520.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Leptidae amurensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Pieridae</td>\n",
       "      <td>Pieris</td>\n",
       "      <td>brassicae</td>\n",
       "      <td>360</td>\n",
       "      <td>450</td>\n",
       "      <td>NaN</td>\n",
       "      <td>560.0</td>\n",
       "      <td>560.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pieris brassicae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Pieridae</td>\n",
       "      <td>Pieris</td>\n",
       "      <td>rapae crucivora</td>\n",
       "      <td>360</td>\n",
       "      <td>440</td>\n",
       "      <td>460</td>\n",
       "      <td>560.0</td>\n",
       "      <td>620.0</td>\n",
       "      <td>640.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pieris rapae crucivora</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Family     Genus          Species First Seco Third  Fourth  Fifth  \\\n",
       "kooi_id                                                                        \n",
       "0        Pieridae    Colias            erate   360  440   480   580.0  660.0   \n",
       "1        Pieridae    Colias            erate   360  430   460   580.0  640.0   \n",
       "2        Pieridae  Leptidae        amurensis   360  440   450   520.0    NaN   \n",
       "3        Pieridae    Pieris        brassicae   360  450   NaN   560.0  560.0   \n",
       "4        Pieridae    Pieris  rapae crucivora   360  440   460   560.0  620.0   \n",
       "\n",
       "         Sixth Remaining            Full_Species  \n",
       "kooi_id                                           \n",
       "0          NaN       NaN            Colias erate  \n",
       "1        620.0       660            Colias erate  \n",
       "2          NaN       NaN      Leptidae amurensis  \n",
       "3          NaN       NaN        Pieris brassicae  \n",
       "4        640.0       NaN  Pieris rapae crucivora  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kooi_insect_db_file = f'{report_dir}/j_van_der_kooi/j_van_der_kooi_2021_main_table.csv'\n",
    "kooi_df = pd.read_csv(kooi_insect_db_file, index_col=0)\n",
    "kooi_df['Full_Species'] = (kooi_df['Genus'] + ' ' + kooi_df['Species']).to_list()\n",
    "df_list.append(caves_df)\n",
    "kooi_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Need a way to add/collapse the different lmax columns into a central column, making sure the corresponding species information is retained."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color=#c994c7>Merge Accessory Lambda Max Databases</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Call the function to merge all the species, lambdamax, and potential accession information into one dataframe\n",
    "# merged_df = merge_accessory_dbs(df_list, report_dir)\n",
    "#merged_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#merged_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color=#c994c7>All unique species names have been extracted from accessory databases. Now we iteratively query NCBI for opsins from each species.</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4098"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(species_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1273"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_list = list(set(species_list))\n",
    "len(species_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color=#c51b8a>Run NCBI Query Mining Process</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating Job Directory\n",
      "\n",
      "Saving Species Query List to Text\n",
      "\n",
      "Constructing Taxon Dictionary, Including Species Synonyms\n",
      "\n",
      "Existing Taxon Dictionary Found! One Moment While We Update It...\n",
      "\n",
      "Taxon Dictionary Complete!\n",
      "\n",
      "Starting Queries to NCBI for Opsin Sequences\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0% (0 of 10) |                         | Elapsed Time: 0:00:00 ETA:  --:--:--\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here is the query ID return list:\n",
      "['2154371071', '2154368583', '2154368061', '2154368053', '2154366825', '2154366823', '2154366821', '2154359214', '2154359197', '2154359195', '2154359192', '2154359184', '2154359181', '2154359180', '2154359175', '2154355812', '2154351657', '2154351655', '2154351652', '2154351649', '2154345030', '2154345027', '2154344058', '2154344036', '2154341425', '2154338363', '2154325160', '2154325154', '2154318689', '2154293566', '2154293562', '2154293558', '2154289014', '2154289010', '2154284796', '2154267127', '2154260939']\n",
      "\n",
      "Fetching record 2154371071 from cache.\n",
      "Fetching record 2154368583 from cache.\n",
      "Fetching record 2154368061 from cache.\n",
      "Fetching record 2154368053 from cache.\n",
      "Fetching record 2154366825 from cache.\n",
      "Fetching record 2154366823 from cache.\n",
      "Fetching record 2154366821 from cache.\n",
      "Fetching record 2154359214 from cache.\n",
      "Fetching record 2154359197 from cache.\n",
      "Fetching record 2154359195 from cache.\n",
      "Fetching record 2154359192 from cache.\n",
      "Fetching record 2154359184 from cache.\n",
      "Fetching record 2154359181 from cache.\n",
      "Fetching record 2154359180 from cache.\n",
      "Fetching record 2154359175 from cache.\n",
      "Fetching record 2154355812 from cache.\n",
      "Fetching record 2154351657 from cache.\n",
      "Fetching record 2154351655 from cache.\n",
      "Fetching record 2154351652 from cache.\n",
      "Fetching record 2154351649 from cache.\n",
      "Fetching record 2154345030 from cache.\n",
      "Fetching record 2154345027 from cache.\n",
      "Fetching record 2154344058 from cache.\n",
      "Fetching record 2154344036 from cache.\n",
      "Fetching record 2154341425 from cache.\n",
      "Fetching record 2154338363 from cache.\n",
      "Fetching record 2154325160 from cache.\n",
      "Fetching record 2154325154 from cache.\n",
      "Fetching record 2154318689 from cache.\n",
      "Fetching record 2154293566 from cache.\n",
      "Fetching record 2154293562 from cache.\n",
      "Fetching record 2154293558 from cache.\n",
      "Fetching record 2154289014 from cache.\n",
      "Fetching record 2154289010 from cache.\n",
      "Fetching record 2154284796 from cache.\n",
      "Fetching record 2154267127 from cache.\n",
      "Fetching record 2154260939 from cache.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 10% (1 of 10) |##                       | Elapsed Time: 0:00:03 ETA:   0:00:27\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here is the query ID return list:\n",
      "[]\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 20% (2 of 10) |#####                    | Elapsed Time: 0:00:06 ETA:   0:00:24\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here is the query ID return list:\n",
      "[]\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 30% (3 of 10) |#######                  | Elapsed Time: 0:00:08 ETA:   0:00:20\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here is the query ID return list:\n",
      "[]\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 40% (4 of 10) |##########               | Elapsed Time: 0:00:11 ETA:   0:00:16\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here is the query ID return list:\n",
      "[]\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 50% (5 of 10) |############             | Elapsed Time: 0:00:14 ETA:   0:00:14\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here is the query ID return list:\n",
      "[]\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 60% (6 of 10) |###############          | Elapsed Time: 0:00:16 ETA:   0:00:10\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here is the query ID return list:\n",
      "[]\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 70% (7 of 10) |#################        | Elapsed Time: 0:00:19 ETA:   0:00:08\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here is the query ID return list:\n",
      "[]\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 80% (8 of 10) |####################     | Elapsed Time: 0:00:21 ETA:   0:00:05\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here is the query ID return list:\n",
      "[]\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 90% (9 of 10) |######################   | Elapsed Time: 0:00:24 ETA:   0:00:02\n",
      "100% (10 of 10) |########################| Elapsed Time: 0:00:24 Time:  0:00:24\n",
      "100% (10 of 10) |########################| Elapsed Time: 0:00:24 Time:  0:00:24\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Here is the query ID return list:\n",
      "[]\n",
      "\n",
      "NCBI Queries Complete!\n",
      "Now Extracting and Formatting Results For DataFrame...\n",
      "\n",
      "DataFrame Formatted and Saved to CSV file for future use :)\n",
      "\n",
      "FASTA File Saved...\n",
      "\n",
      "Saving txt file with names of species that retrieved no results for opsins...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ncbi_query_df, query_report_dir = ncbi_fetch_opsins(email=email, job_label='mnm_on_all_dbs', out='mnm_on_all_dbs', species_list=species_list[0:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'STOP' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[20], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[43mSTOP\u001b[49m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'STOP' is not defined"
     ]
    }
   ],
   "source": [
    "STOP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1957, 12)"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_query_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color=#c51b8a>Load Accessory Opsin Sequence Databases</font> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color=#c994c7>Load Previous MnM Data</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "#species_list = []\n",
    "#with open(f'{query_report_dir}/species_queried.txt', \"r\") as f:\n",
    "#    lines = f.readlines()\n",
    "#    for line in lines:\n",
    "#        species_list.append(line.replace('\\n',''))\n",
    "#len(species_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Subphylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>Full_Species</th>\n",
       "      <th>DNA</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Gene_Description</th>\n",
       "      <th>Species_Synonym_Used</th>\n",
       "      <th>Prot_Len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XM_024085765.2</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexapoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Bicyclus</td>\n",
       "      <td>anynana</td>\n",
       "      <td>Bicyclus anynana</td>\n",
       "      <td>ACACAATTCATTTAACGTGTTCAGAATAGATGCACCACCGCATTAG...</td>\n",
       "      <td>MATQSCFWSDIFDEQSRANLNTGFVVTGAVSTLLGSWLNLALLLTN...</td>\n",
       "      <td>PREDICTED: Bicyclus anynana rhodopsin (LOC1120...</td>\n",
       "      <td>NA</td>\n",
       "      <td>313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XM_024089198.2</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexapoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Bicyclus</td>\n",
       "      <td>anynana</td>\n",
       "      <td>Bicyclus anynana</td>\n",
       "      <td>GTTCGGGATGACAGATATACGCTCGAGAATAACATTTTGCTACGGG...</td>\n",
       "      <td>MSPATKIIVIIFLIDIKCKNVDTLPEVKAELWSFKNNLKKSVVNHF...</td>\n",
       "      <td>PREDICTED: Bicyclus anynana opsin, ultraviolet...</td>\n",
       "      <td>NA</td>\n",
       "      <td>572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AY918895.2</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexapoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Bicyclus</td>\n",
       "      <td>anynana</td>\n",
       "      <td>Bicyclus anynana</td>\n",
       "      <td>TTTGGTATTAACCATCTCGGGGCGACCTATCCACCTGTAACCCTTT...</td>\n",
       "      <td>MAITSMDPGPGIAALQAWGGHAAAYGSNETVVDKVLPDMLHLVDPH...</td>\n",
       "      <td>Bicyclus anynana long wavelength sensitive ops...</td>\n",
       "      <td>NA</td>\n",
       "      <td>380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XM_061286956.1</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Craniata</td>\n",
       "      <td>Actinopteri</td>\n",
       "      <td>Syngnathus</td>\n",
       "      <td>typhle</td>\n",
       "      <td>Syngnathus typhle</td>\n",
       "      <td>ACCGCACGCGTGGATTACATTGCAGCGTGGAGGAAATGGGGAAAAA...</td>\n",
       "      <td>MDARMFSGQVALNLSGSNLSDDPELAEPSVEVEARLSPTGFVVLSV...</td>\n",
       "      <td>PREDICTED: Syngnathus typhle teleost multiple ...</td>\n",
       "      <td>NA</td>\n",
       "      <td>381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XM_061286026.1</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Craniata</td>\n",
       "      <td>Actinopteri</td>\n",
       "      <td>Syngnathus</td>\n",
       "      <td>typhle</td>\n",
       "      <td>Syngnathus typhle</td>\n",
       "      <td>GGACGCGCGCTCCAGCCGCACATCGAAGCCACCACAATGACCGAAA...</td>\n",
       "      <td>MNPYNETRAEHNLFAFGTYKLLAVTIGTIGVFGFCNNVLVILLYCK...</td>\n",
       "      <td>PREDICTED: Syngnathus typhle opsin 3 (opn3), mRNA</td>\n",
       "      <td>NA</td>\n",
       "      <td>381</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Accession      Phylum Subphylum        Class       Genus  Species  \\\n",
       "0  XM_024085765.2  Arthropoda  Hexapoda      Insecta    Bicyclus  anynana   \n",
       "1  XM_024089198.2  Arthropoda  Hexapoda      Insecta    Bicyclus  anynana   \n",
       "2      AY918895.2  Arthropoda  Hexapoda      Insecta    Bicyclus  anynana   \n",
       "3  XM_061286956.1    Chordata  Craniata  Actinopteri  Syngnathus   typhle   \n",
       "4  XM_061286026.1    Chordata  Craniata  Actinopteri  Syngnathus   typhle   \n",
       "\n",
       "        Full_Species                                                DNA  \\\n",
       "0   Bicyclus anynana  ACACAATTCATTTAACGTGTTCAGAATAGATGCACCACCGCATTAG...   \n",
       "1   Bicyclus anynana  GTTCGGGATGACAGATATACGCTCGAGAATAACATTTTGCTACGGG...   \n",
       "2   Bicyclus anynana  TTTGGTATTAACCATCTCGGGGCGACCTATCCACCTGTAACCCTTT...   \n",
       "3  Syngnathus typhle  ACCGCACGCGTGGATTACATTGCAGCGTGGAGGAAATGGGGAAAAA...   \n",
       "4  Syngnathus typhle  GGACGCGCGCTCCAGCCGCACATCGAAGCCACCACAATGACCGAAA...   \n",
       "\n",
       "                                             Protein  \\\n",
       "0  MATQSCFWSDIFDEQSRANLNTGFVVTGAVSTLLGSWLNLALLLTN...   \n",
       "1  MSPATKIIVIIFLIDIKCKNVDTLPEVKAELWSFKNNLKKSVVNHF...   \n",
       "2  MAITSMDPGPGIAALQAWGGHAAAYGSNETVVDKVLPDMLHLVDPH...   \n",
       "3  MDARMFSGQVALNLSGSNLSDDPELAEPSVEVEARLSPTGFVVLSV...   \n",
       "4  MNPYNETRAEHNLFAFGTYKLLAVTIGTIGVFGFCNNVLVILLYCK...   \n",
       "\n",
       "                                    Gene_Description Species_Synonym_Used  \\\n",
       "0  PREDICTED: Bicyclus anynana rhodopsin (LOC1120...                   NA   \n",
       "1  PREDICTED: Bicyclus anynana opsin, ultraviolet...                   NA   \n",
       "2  Bicyclus anynana long wavelength sensitive ops...                   NA   \n",
       "3  PREDICTED: Syngnathus typhle teleost multiple ...                   NA   \n",
       "4  PREDICTED: Syngnathus typhle opsin 3 (opn3), mRNA                   NA   \n",
       "\n",
       "   Prot_Len  \n",
       "0       313  \n",
       "1       572  \n",
       "2       380  \n",
       "3       381  \n",
       "4       381  "
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#query_report_dir = \"mnm_data/mnm_on_mnm_on_all_dbs_2025-01-22_20-33-16 - Copy\"\n",
    "#ncbi_query_file = f'{query_report_dir}/mnm_on_all_dbs_ncbi_q_data_cleaned.csv'\n",
    "#ncbi_query_file = f'./{query_report_dir}/mnm_on_all_dbs_ncbi_q_data_cleaned.csv'\n",
    "#ncbi_query_df = pd.read_csv(ncbi_query_file)\n",
    "ncbi_sp_list = ncbi_query_df['Full_Species'].to_list()\n",
    "ncbi_prot_list = ncbi_query_df['Protein'].to_list()\n",
    "ncbi_query_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "report_dir2 = './data_sources/seqs'\n",
    "sequence_list = []\n",
    "acc_list = []\n",
    "db_sp_list = []\n",
    "prot_descriptions = []\n",
    "source_list = []"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color=#c994c7>Bilaterian Animal Opsins: Ramirez et al 2016</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>aligned_seqs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KX714605</td>\n",
       "      <td>Acanthopleura granulata</td>\n",
       "      <td>xenopsin</td>\n",
       "      <td>----------------------------------------------...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KX714606</td>\n",
       "      <td>Acanthopleura granulata</td>\n",
       "      <td>canonical_r-opsin</td>\n",
       "      <td>---------YII-------------------------------GVY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_Acroporadigitifera</td>\n",
       "      <td>Acropora digitifera</td>\n",
       "      <td>58percent_match_PREDICTED visual pigment-like ...</td>\n",
       "      <td>---------Y------------------------------------...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XP_015763203</td>\n",
       "      <td>Acropora digitifera</td>\n",
       "      <td>na</td>\n",
       "      <td>---------IVY-------------------------------VVI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XP_015773304</td>\n",
       "      <td>Acropora digitifera</td>\n",
       "      <td>PREDICTED: melanopsin-like</td>\n",
       "      <td>---------IAY-------------------------------GVV...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_ID                 Organism  \\\n",
       "0              KX714605  Acanthopleura granulata   \n",
       "1              KX714606  Acanthopleura granulata   \n",
       "2  1_Acroporadigitifera      Acropora digitifera   \n",
       "3          XP_015763203      Acropora digitifera   \n",
       "4          XP_015773304      Acropora digitifera   \n",
       "\n",
       "                                       Protein names  \\\n",
       "0                                           xenopsin   \n",
       "1                                  canonical_r-opsin   \n",
       "2  58percent_match_PREDICTED visual pigment-like ...   \n",
       "3                                                 na   \n",
       "4                         PREDICTED: melanopsin-like   \n",
       "\n",
       "                                        aligned_seqs  \n",
       "0  ----------------------------------------------...  \n",
       "1  ---------YII-------------------------------GVY...  \n",
       "2  ---------Y------------------------------------...  \n",
       "3  ---------IVY-------------------------------VVI...  \n",
       "4  ---------IAY-------------------------------GVV...  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ramirez_file = f'{report_dir2}/bilaterian_animal_ops/Supplemental_Table_T1_cleaned.csv'\n",
    "ramirez_df = pd.read_csv(ramirez_file)\n",
    "ramirez_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(768, 4)"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ramirez_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\safra\\AppData\\Local\\Temp\\ipykernel_29724\\4088448753.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ramirez_df_filtered['raw_seqs'] = ramirez_df_filtered['aligned_seqs'].str.replace('-', '')\n",
      "C:\\Users\\safra\\AppData\\Local\\Temp\\ipykernel_29724\\4088448753.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ramirez_df_filtered['prot_len'] = prot_len_list\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>aligned_seqs</th>\n",
       "      <th>raw_seqs</th>\n",
       "      <th>prot_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>U3IU26</td>\n",
       "      <td>Anas platyrhynchos</td>\n",
       "      <td>Uncharacterized protein</td>\n",
       "      <td>TQTEH----NIV-------------------------------AAY...</td>\n",
       "      <td>TQTEHNIVAAYLITAGVISIFSNIVVLGIFVKYKELRTATNAIIIN...</td>\n",
       "      <td>340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0R4IEG0</td>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>Novopsin-10</td>\n",
       "      <td>SPSAD----LSI-------------------------------AVF...</td>\n",
       "      <td>SPSADLSIAVFLIITGVVSVFGNGLVLLVYGQRRKKLRAHELMTIN...</td>\n",
       "      <td>378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q80XL3</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Rrh protein</td>\n",
       "      <td>SRTEH----SVI-------------------------------AAY...</td>\n",
       "      <td>SRTEHSVIAAYLIVAGITSILSNVVVLGIFIKYKELRTPTNAVIIN...</td>\n",
       "      <td>331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M3YAY1</td>\n",
       "      <td>Mustela putorius furo</td>\n",
       "      <td>Uncharacterized protein</td>\n",
       "      <td>SQTEH----NIV-------------------------------AAY...</td>\n",
       "      <td>SQTEHNIVAAYLITAGIISIFSNLIVLGIFIKYKELRTPTNAIIIN...</td>\n",
       "      <td>328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G1PJY3</td>\n",
       "      <td>Myotis lucifugus</td>\n",
       "      <td>Uncharacterized protein</td>\n",
       "      <td>SQTEH----NIV-------------------------------ATY...</td>\n",
       "      <td>SQTEHNIVATYLIMAGMISLLSNIIVLGIFITYKELRTPTNAIIIN...</td>\n",
       "      <td>331</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_ID               Organism            Protein names  \\\n",
       "0      U3IU26     Anas platyrhynchos  Uncharacterized protein   \n",
       "1  A0A0R4IEG0            Danio rerio              Novopsin-10   \n",
       "2      Q80XL3           Mus musculus              Rrh protein   \n",
       "3      M3YAY1  Mustela putorius furo  Uncharacterized protein   \n",
       "4      G1PJY3       Myotis lucifugus  Uncharacterized protein   \n",
       "\n",
       "                                        aligned_seqs  \\\n",
       "0  TQTEH----NIV-------------------------------AAY...   \n",
       "1  SPSAD----LSI-------------------------------AVF...   \n",
       "2  SRTEH----SVI-------------------------------AAY...   \n",
       "3  SQTEH----NIV-------------------------------AAY...   \n",
       "4  SQTEH----NIV-------------------------------ATY...   \n",
       "\n",
       "                                            raw_seqs  prot_len  \n",
       "0  TQTEHNIVAAYLITAGVISIFSNIVVLGIFVKYKELRTATNAIIIN...       340  \n",
       "1  SPSADLSIAVFLIITGVVSVFGNGLVLLVYGQRRKKLRAHELMTIN...       378  \n",
       "2  SRTEHSVIAAYLIVAGITSILSNVVVLGIFIKYKELRTPTNAVIIN...       331  \n",
       "3  SQTEHNIVAAYLITAGIISIFSNLIVLGIFIKYKELRTPTNAIIIN...       328  \n",
       "4  SQTEHNIVATYLIMAGMISLLSNIIVLGIFITYKELRTPTNAIIIN...       331  "
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ramirez_df_filtered = ramirez_df[ramirez_df['Organism'].isin(species_list)]\n",
    "ramirez_df_filtered['raw_seqs'] = ramirez_df_filtered['aligned_seqs'].str.replace('-', '')\n",
    "prot_len_list = []\n",
    "for prot in ramirez_df_filtered['raw_seqs']:\n",
    "    prot_len = len(prot)\n",
    "    prot_len_list.append(prot_len)\n",
    "ramirez_df_filtered['prot_len'] = prot_len_list\n",
    "ramirez_df_filtered = ramirez_df_filtered[ramirez_df_filtered['prot_len']>320].reset_index(drop=True)\n",
    "ramirez_df_filtered.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6, 6)"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ramirez_df_filtered.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(ramirez_df_filtered['Organism'].to_list()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6, 7)"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ramirez_df_filtered2 = ramirez_df_filtered[~ramirez_df_filtered['raw_seqs'].isin(ncbi_prot_list)]\n",
    "ramirez_df_filtered2 = ramirez_df_filtered2.reset_index(drop=True)\n",
    "ramirez_df_filtered2['source'] = 'Ramirez et al 2016'\n",
    "ramirez_df_filtered2.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ramirez_df_filtered2.to_csv(f'{report_dir2}/bilaterian_animal_ops/Supplemental_Table_T1_filtered.csv')\n",
    "len(set(ramirez_df_filtered2['raw_seqs'].to_list()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequence_list += ramirez_df_filtered2['raw_seqs'].to_list()\n",
    "acc_list += ramirez_df_filtered2['gene_ID'].to_list()\n",
    "db_sp_list += ramirez_df_filtered2['Organism'].to_list()\n",
    "prot_descriptions += ramirez_df_filtered2['Protein names'].to_list()\n",
    "source_list += ramirez_df_filtered2['source'].to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color=#c994c7>Crustacean Conundrums: Palecanda et al 2022</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus species</th>\n",
       "      <th>Opsin Type</th>\n",
       "      <th>NCBI Accession #</th>\n",
       "      <th>Transcript ID</th>\n",
       "      <th>Untrimmed Nucleotide Sequence</th>\n",
       "      <th>Protein Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Malacostraca</td>\n",
       "      <td>Amphipoda</td>\n",
       "      <td>Corophiidae</td>\n",
       "      <td>Grandidierella japonica</td>\n",
       "      <td>LWS</td>\n",
       "      <td>BK059768</td>\n",
       "      <td>AMP|GrJap|LWS|IACS01070746.1</td>\n",
       "      <td>ACTGAGAAATATCTACGGAAAAATGTCTTGGAATAGCCCAATGAAC...</td>\n",
       "      <td>LPEGMASSNPFGNYTVVDAAPKEILHMVDPHWYQFPPMNPLWYGLL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Malacostraca</td>\n",
       "      <td>Amphipoda</td>\n",
       "      <td>Corophiidae</td>\n",
       "      <td>Grandidierella japonica</td>\n",
       "      <td>MWS1</td>\n",
       "      <td>BK059769</td>\n",
       "      <td>AMP|GrJap|MWS|IACS01055269.1</td>\n",
       "      <td>GGAGTCACCCAGGTTGACCTGGTTCCGGATTACATGAAGGACATGA...</td>\n",
       "      <td>GVTQVDLVPDYMKDMIHPHWSNYPPVNPMWHHLLGLVYIIIGSCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Malacostraca</td>\n",
       "      <td>Amphipoda</td>\n",
       "      <td>Corophiidae</td>\n",
       "      <td>Grandidierella japonica</td>\n",
       "      <td>NEUR</td>\n",
       "      <td>BK059770</td>\n",
       "      <td>AMP|GrJap|NEUR|IACS01028177.1</td>\n",
       "      <td>AGTGTACGGCTACGTATGCTTCGCTCTGGGTATCGTCAATATTCTC...</td>\n",
       "      <td>VYGYVCFALGIVNILSIYALAVVRYLKTCHYRNIGSKIDRKEVSIV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Malacostraca</td>\n",
       "      <td>Amphipoda</td>\n",
       "      <td>Gammaridae</td>\n",
       "      <td>Echinogammarus berilloni</td>\n",
       "      <td>LWS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AMP|EcBer|LWS|GHCT01076719.1</td>\n",
       "      <td>CGTGGCTTCAACGCGCTGTGTAGACCAACAGGACCTCCGCCTCACT...</td>\n",
       "      <td>LPEGVVSSNPFGNYTVVDAAPKELLPLIDSHWYQYPPLNPMWYGLL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Malacostraca</td>\n",
       "      <td>Amphipoda</td>\n",
       "      <td>Gammaridae</td>\n",
       "      <td>Echinogammarus berilloni</td>\n",
       "      <td>LWS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AMP|EcBer|LWS|GHCT01076725.1</td>\n",
       "      <td>CTCAATATATAATTTGAGCGTCCAGAAGATTCTGTCTCAGTGAGCC...</td>\n",
       "      <td>APEGAVSTNPFGNYTVVDTAAKDILHMISPHWYQFPPLNPMWYGLL...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Class      Order       Family             Genus species Opsin Type  \\\n",
       "0  Malacostraca  Amphipoda  Corophiidae   Grandidierella japonica        LWS   \n",
       "1  Malacostraca  Amphipoda  Corophiidae   Grandidierella japonica       MWS1   \n",
       "2  Malacostraca  Amphipoda  Corophiidae   Grandidierella japonica       NEUR   \n",
       "3  Malacostraca  Amphipoda   Gammaridae  Echinogammarus berilloni        LWS   \n",
       "4  Malacostraca  Amphipoda   Gammaridae  Echinogammarus berilloni        LWS   \n",
       "\n",
       "  NCBI Accession #                  Transcript ID  \\\n",
       "0         BK059768   AMP|GrJap|LWS|IACS01070746.1   \n",
       "1         BK059769   AMP|GrJap|MWS|IACS01055269.1   \n",
       "2         BK059770  AMP|GrJap|NEUR|IACS01028177.1   \n",
       "3              NaN   AMP|EcBer|LWS|GHCT01076719.1   \n",
       "4              NaN   AMP|EcBer|LWS|GHCT01076725.1   \n",
       "\n",
       "                       Untrimmed Nucleotide Sequence  \\\n",
       "0  ACTGAGAAATATCTACGGAAAAATGTCTTGGAATAGCCCAATGAAC...   \n",
       "1  GGAGTCACCCAGGTTGACCTGGTTCCGGATTACATGAAGGACATGA...   \n",
       "2  AGTGTACGGCTACGTATGCTTCGCTCTGGGTATCGTCAATATTCTC...   \n",
       "3  CGTGGCTTCAACGCGCTGTGTAGACCAACAGGACCTCCGCCTCACT...   \n",
       "4  CTCAATATATAATTTGAGCGTCCAGAAGATTCTGTCTCAGTGAGCC...   \n",
       "\n",
       "                                    Protein Sequence  \n",
       "0  LPEGMASSNPFGNYTVVDAAPKEILHMVDPHWYQFPPMNPLWYGLL...  \n",
       "1  GVTQVDLVPDYMKDMIHPHWSNYPPVNPMWHHLLGLVYIIIGSCAT...  \n",
       "2  VYGYVCFALGIVNILSIYALAVVRYLKTCHYRNIGSKIDRKEVSIV...  \n",
       "3  LPEGVVSSNPFGNYTVVDAAPKELLPLIDSHWYQYPPLNPMWYGLL...  \n",
       "4  APEGAVSTNPFGNYTVVDTAAKDILHMISPHWYQFPPLNPMWYGLL...  "
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "palecanda_file = f'{report_dir2}/crustacean_conundrums/rstb20210289_si_003.csv'\n",
    "palecanda_df = pd.read_csv(palecanda_file)\n",
    "palecanda_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(628, 9)"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "palecanda_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus species</th>\n",
       "      <th>Opsin Type</th>\n",
       "      <th>NCBI Accession #</th>\n",
       "      <th>Transcript ID</th>\n",
       "      <th>Untrimmed Nucleotide Sequence</th>\n",
       "      <th>Protein Sequence</th>\n",
       "      <th>prot_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Malacostraca</td>\n",
       "      <td>Amphipoda</td>\n",
       "      <td>Talitridae</td>\n",
       "      <td>Talitrus saltator</td>\n",
       "      <td>LWS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AMP|TeSal|LWS|GDUJ01056749.1</td>\n",
       "      <td>TCACGTGGCCTAGCGCGCTGTGTAGACCGAATCCAACTCCTTGTCT...</td>\n",
       "      <td>HPEGMVSSNPFGNFTVVDAAPKELLHMVDPHWYQYPPLNPLWYGLL...</td>\n",
       "      <td>336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Malacostraca</td>\n",
       "      <td>Amphipoda</td>\n",
       "      <td>Talitridae</td>\n",
       "      <td>Talitrus saltator</td>\n",
       "      <td>LWS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AMP|TeSal|LWS|GDUJ01056750.1</td>\n",
       "      <td>CTCACAACGAATCCAACTCATCAGTAATAATGAGGTGGAGGCGCGT...</td>\n",
       "      <td>LPDSGYISTNPFGNFTVVDSAPKEILHMIDPHWYQFPPLNPMWYGL...</td>\n",
       "      <td>337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Malacostraca</td>\n",
       "      <td>Amphipoda</td>\n",
       "      <td>Talitridae</td>\n",
       "      <td>Talitrus saltator</td>\n",
       "      <td>MWS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AMP|TeSal|MWS|GDUJ01058424.1</td>\n",
       "      <td>TGGAGGGCTTCTTGGGGCCTCCCAGCTGCATTCATGTCCCTAACAC...</td>\n",
       "      <td>SYGSPSSSALTGANQFAFGYPPGVSVVDIVPPHMKDLIHPHWSNFP...</td>\n",
       "      <td>373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Malacostraca</td>\n",
       "      <td>Amphipoda</td>\n",
       "      <td>Talitridae</td>\n",
       "      <td>Talitrus saltator</td>\n",
       "      <td>NEUR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AMP|TeSal|NEUR|GDUJ01037444.1</td>\n",
       "      <td>ATGGGTTAAAAATCAGCGCTGCATTGACGAGAAGTTTGACTCATCT...</td>\n",
       "      <td>VVQTRPIFIISTRFTLHFDTEWLLACMTAGVVALVGNSASVAMFWR...</td>\n",
       "      <td>330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Malacostraca</td>\n",
       "      <td>Decapoda</td>\n",
       "      <td>Cambaridae</td>\n",
       "      <td>Procambarus clarkii</td>\n",
       "      <td>LWS</td>\n",
       "      <td>ALJ26467</td>\n",
       "      <td>DEC|Proclar|LWS|ALJ26467</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SWSNQPAMDDYGLPSSNPYGNFTVVDMAPKDILHMIHPHWYQYPPM...</td>\n",
       "      <td>363</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Class      Order      Family        Genus species Opsin Type  \\\n",
       "0  Malacostraca  Amphipoda  Talitridae    Talitrus saltator        LWS   \n",
       "1  Malacostraca  Amphipoda  Talitridae    Talitrus saltator        LWS   \n",
       "2  Malacostraca  Amphipoda  Talitridae    Talitrus saltator       MWS1   \n",
       "3  Malacostraca  Amphipoda  Talitridae    Talitrus saltator       NEUR   \n",
       "4  Malacostraca   Decapoda  Cambaridae  Procambarus clarkii        LWS   \n",
       "\n",
       "  NCBI Accession #                  Transcript ID  \\\n",
       "0              NaN   AMP|TeSal|LWS|GDUJ01056749.1   \n",
       "1              NaN   AMP|TeSal|LWS|GDUJ01056750.1   \n",
       "2              NaN   AMP|TeSal|MWS|GDUJ01058424.1   \n",
       "3              NaN  AMP|TeSal|NEUR|GDUJ01037444.1   \n",
       "4         ALJ26467       DEC|Proclar|LWS|ALJ26467   \n",
       "\n",
       "                       Untrimmed Nucleotide Sequence  \\\n",
       "0  TCACGTGGCCTAGCGCGCTGTGTAGACCGAATCCAACTCCTTGTCT...   \n",
       "1  CTCACAACGAATCCAACTCATCAGTAATAATGAGGTGGAGGCGCGT...   \n",
       "2  TGGAGGGCTTCTTGGGGCCTCCCAGCTGCATTCATGTCCCTAACAC...   \n",
       "3  ATGGGTTAAAAATCAGCGCTGCATTGACGAGAAGTTTGACTCATCT...   \n",
       "4                                                NaN   \n",
       "\n",
       "                                    Protein Sequence  prot_len  \n",
       "0  HPEGMVSSNPFGNFTVVDAAPKELLHMVDPHWYQYPPLNPLWYGLL...       336  \n",
       "1  LPDSGYISTNPFGNFTVVDSAPKEILHMIDPHWYQFPPLNPMWYGL...       337  \n",
       "2  SYGSPSSSALTGANQFAFGYPPGVSVVDIVPPHMKDLIHPHWSNFP...       373  \n",
       "3  VVQTRPIFIISTRFTLHFDTEWLLACMTAGVVALVGNSASVAMFWR...       330  \n",
       "4  SWSNQPAMDDYGLPSSNPYGNFTVVDMAPKDILHMIHPHWYQYPPM...       363  "
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "palecanda_df_filtered = palecanda_df[palecanda_df['Genus species'].isin(species_list)].reset_index(drop=True)\n",
    "#palecanda_df_filtered['raw_seqs'] = palecanda_df_filtered['aligned_seqs'].str.replace('-', '')\n",
    "prot_len_list = []\n",
    "for prot in palecanda_df_filtered['Protein Sequence']:\n",
    "    prot_len = len(prot)\n",
    "    prot_len_list.append(prot_len)\n",
    "palecanda_df_filtered['prot_len'] = prot_len_list\n",
    "palecanda_df_filtered = palecanda_df_filtered[palecanda_df_filtered['prot_len']>320].reset_index(drop=True)\n",
    "\n",
    "palecanda_df_filtered.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(84, 10)"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "palecanda_df_filtered.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(palecanda_df_filtered['Genus species'].to_list()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(84, 11)"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "palecanda_df_filtered2 = palecanda_df_filtered[~palecanda_df_filtered['Protein Sequence'].isin(ncbi_prot_list)]\n",
    "palecanda_df_filtered2=palecanda_df_filtered2.reset_index(drop=True)\n",
    "palecanda_df_filtered2['source'] = 'Palecanda et al 2022'\n",
    "palecanda_df_filtered2.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "84"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "palecanda_df_filtered2.to_csv(f'{report_dir2}/crustacean_conundrums/palecanda_2022_filtered_seq_data.csv')\n",
    "len(set(palecanda_df_filtered2['Protein Sequence'].to_list()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequence_list += palecanda_df_filtered2['Protein Sequence'].to_list()\n",
    "acc_list += palecanda_df_filtered2['Transcript ID'].to_list()\n",
    "db_sp_list += palecanda_df_filtered2['Genus species'].to_list()\n",
    "prot_descriptions += palecanda_df_filtered2['Opsin Type'].to_list()\n",
    "source_list += palecanda_df_filtered2['source'].to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color=#c994c7>Ray-Finned Fish Opsin DB: Policarpo et al 2024</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "ray_finned_fasta_file = f'{report_dir2}/ray_finned_fish/mined_alignments/Complete_opsins.aln.txt'\n",
    "ray_finned_df = fasta_to_dataframe(ray_finned_fasta_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species_name</th>\n",
       "      <th>opsin_type</th>\n",
       "      <th>accession</th>\n",
       "      <th>aln_sequence</th>\n",
       "      <th>sequence</th>\n",
       "      <th>seq_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Lepisosteus oculatus</td>\n",
       "      <td>va</td>\n",
       "      <td>CM001408.1-11749388-11761955</td>\n",
       "      <td>MDSFRVSVNG-VV--------------------------YTEAAE-...</td>\n",
       "      <td>MDSFRVSVNGVVYTEAAEILKPGDPFSGPIENIAPWNFKFLAALMF...</td>\n",
       "      <td>386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Lepisosteus oculatus</td>\n",
       "      <td>pinopsin</td>\n",
       "      <td>CM001425.1-1263197-1266918</td>\n",
       "      <td>MPILVNSSAA-FP--------------------------L--EKN-...</td>\n",
       "      <td>MPILVNSSAAFPLEKNSTPGPFDGPQWHQAPRSTYLMVAVLMGTVV...</td>\n",
       "      <td>357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Lepisosteus oculatus</td>\n",
       "      <td>sws1</td>\n",
       "      <td>CM001411.1-10398717-10404663</td>\n",
       "      <td>MAGTEE--FY-L---------------------------F----E-...</td>\n",
       "      <td>MAGTEEFYLFENISSVGPWDGPQYHIAPKWAFYFQTIFMGLVFFWG...</td>\n",
       "      <td>346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Lepisosteus oculatus</td>\n",
       "      <td>rrh</td>\n",
       "      <td>CM001407.1-48235198-48246929</td>\n",
       "      <td>MENFSRIPIT-KV--------------------------LDNETS-...</td>\n",
       "      <td>MENFSRIPITKVLDNETSNLSQAVDTSEHSVFSQTEHNIVAAYLIT...</td>\n",
       "      <td>343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Lepisosteus oculatus</td>\n",
       "      <td>opn5</td>\n",
       "      <td>CM001404.1-37427058-37472360</td>\n",
       "      <td>MAQTDNGTSHNVP--------------------------HYL----...</td>\n",
       "      <td>MAQTDNGTSHNVPHYLLSGDPFASKLSKEADFVAAFYICIVGIMST...</td>\n",
       "      <td>352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17336</th>\n",
       "      <td>Callorhinchus milii</td>\n",
       "      <td>pinopsin</td>\n",
       "      <td>XP007894735.2</td>\n",
       "      <td>----MDLPTT-SV--------------------------G--IEN-...</td>\n",
       "      <td>MDLPTTSVGIENGTVGAFDGPQWYFAPKSTYMAVATLMGTVVILAS...</td>\n",
       "      <td>352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17337</th>\n",
       "      <td>Carcharodon carcharias</td>\n",
       "      <td>pinopsin</td>\n",
       "      <td>XP041052689.1</td>\n",
       "      <td>----MGFSTG-SI--------------------------T--THN-...</td>\n",
       "      <td>MGFSTGSITTHNITIGAFHGPQWDVAPRSTYMAVAALMGNVVILAT...</td>\n",
       "      <td>352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17338</th>\n",
       "      <td>Callorhinchus milii</td>\n",
       "      <td>rgr</td>\n",
       "      <td>XM007898380.1</td>\n",
       "      <td>MVTSHP----------------------------------------...</td>\n",
       "      <td>MVTSHPGLEGFTDFEVFGLLVEALVGLLLNGLTLLAFYKIKELRTP...</td>\n",
       "      <td>291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17339</th>\n",
       "      <td>Rhincodon typus</td>\n",
       "      <td>rh1</td>\n",
       "      <td>A0A7T8R2L6</td>\n",
       "      <td>MNGTEGENFY-IP--------------------------MSNKTG-...</td>\n",
       "      <td>MNGTEGENFYIPMSNKTGVVRSPFEYPQYYLAEPWKFSLLAAYMFF...</td>\n",
       "      <td>353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17340</th>\n",
       "      <td>Callorhinchus milii</td>\n",
       "      <td>parapinopsin</td>\n",
       "      <td>V9KWA1</td>\n",
       "      <td>--------------------------------------------G-...</td>\n",
       "      <td>GPGLGPGWPSVRAPLSLVMAVISLSSIVLNSLAIAVVLRFQVLQQP...</td>\n",
       "      <td>312</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17341 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 species_name    opsin_type                     accession  \\\n",
       "0        Lepisosteus oculatus            va  CM001408.1-11749388-11761955   \n",
       "1        Lepisosteus oculatus      pinopsin    CM001425.1-1263197-1266918   \n",
       "2        Lepisosteus oculatus          sws1  CM001411.1-10398717-10404663   \n",
       "3        Lepisosteus oculatus           rrh  CM001407.1-48235198-48246929   \n",
       "4        Lepisosteus oculatus          opn5  CM001404.1-37427058-37472360   \n",
       "...                       ...           ...                           ...   \n",
       "17336     Callorhinchus milii      pinopsin                 XP007894735.2   \n",
       "17337  Carcharodon carcharias      pinopsin                 XP041052689.1   \n",
       "17338     Callorhinchus milii           rgr                 XM007898380.1   \n",
       "17339         Rhincodon typus           rh1                    A0A7T8R2L6   \n",
       "17340     Callorhinchus milii  parapinopsin                        V9KWA1   \n",
       "\n",
       "                                            aln_sequence  \\\n",
       "0      MDSFRVSVNG-VV--------------------------YTEAAE-...   \n",
       "1      MPILVNSSAA-FP--------------------------L--EKN-...   \n",
       "2      MAGTEE--FY-L---------------------------F----E-...   \n",
       "3      MENFSRIPIT-KV--------------------------LDNETS-...   \n",
       "4      MAQTDNGTSHNVP--------------------------HYL----...   \n",
       "...                                                  ...   \n",
       "17336  ----MDLPTT-SV--------------------------G--IEN-...   \n",
       "17337  ----MGFSTG-SI--------------------------T--THN-...   \n",
       "17338  MVTSHP----------------------------------------...   \n",
       "17339  MNGTEGENFY-IP--------------------------MSNKTG-...   \n",
       "17340  --------------------------------------------G-...   \n",
       "\n",
       "                                                sequence  seq_length  \n",
       "0      MDSFRVSVNGVVYTEAAEILKPGDPFSGPIENIAPWNFKFLAALMF...         386  \n",
       "1      MPILVNSSAAFPLEKNSTPGPFDGPQWHQAPRSTYLMVAVLMGTVV...         357  \n",
       "2      MAGTEEFYLFENISSVGPWDGPQYHIAPKWAFYFQTIFMGLVFFWG...         346  \n",
       "3      MENFSRIPITKVLDNETSNLSQAVDTSEHSVFSQTEHNIVAAYLIT...         343  \n",
       "4      MAQTDNGTSHNVPHYLLSGDPFASKLSKEADFVAAFYICIVGIMST...         352  \n",
       "...                                                  ...         ...  \n",
       "17336  MDLPTTSVGIENGTVGAFDGPQWYFAPKSTYMAVATLMGTVVILAS...         352  \n",
       "17337  MGFSTGSITTHNITIGAFHGPQWDVAPRSTYMAVAALMGNVVILAT...         352  \n",
       "17338  MVTSHPGLEGFTDFEVFGLLVEALVGLLLNGLTLLAFYKIKELRTP...         291  \n",
       "17339  MNGTEGENFYIPMSNKTGVVRSPFEYPQYYLAEPWKFSLLAAYMFF...         353  \n",
       "17340  GPGLGPGWPSVRAPLSLVMAVISLSSIVLNSLAIAVVLRFQVLQQP...         312  \n",
       "\n",
       "[17341 rows x 6 columns]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ray_finned_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "ray_finned_df.to_csv(f'{report_dir2}/ray_finned_fish/mined_alignments/extracted_fish_ops.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2494, 6)"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ray_finned_df_filtered = ray_finned_df[ray_finned_df['species_name'].isin(species_list)].reset_index(drop=True)\n",
    "ray_finned_df_filtered.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2457, 6)"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ray_finned_df_filtered2 = ray_finned_df_filtered[~ray_finned_df_filtered['sequence'].isin(ncbi_prot_list)].reset_index(drop=True)\n",
    "ray_finned_df_filtered2.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2135, 7)"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ray_finned_df_filtered3 = ray_finned_df_filtered2[ray_finned_df_filtered2['seq_length'] > 320].reset_index(drop=True)\n",
    "ray_finned_df_filtered3['source'] = 'Policarpo et al 2024'\n",
    "ray_finned_df_filtered3.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "ray_finned_df_filtered3.to_csv(f'{report_dir2}/ray_finned_fish/mined_alignments/extracted_fish_ops_filtered.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequence_list += ray_finned_df_filtered3['sequence'].to_list()\n",
    "acc_list += ray_finned_df_filtered3['accession'].to_list()\n",
    "db_sp_list += ray_finned_df_filtered3['species_name'].to_list()\n",
    "prot_descriptions += ray_finned_df_filtered3['opsin_type'].to_list()\n",
    "source_list += ray_finned_df_filtered3['source'].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "gn_list = []\n",
    "sp_list = []\n",
    "for sp in db_sp_list:\n",
    "#    print(sp)\n",
    "    gn_list.append(sp.split(' ', 1)[0])\n",
    "    sp_list.append(sp.split(' ', 1)[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load our existing taxonomy dictionary and pull relevant taxon info\n",
    "taxon_file = './data_sources/taxonomy/ncbi_taxon_dict.json'\n",
    "if os.path.isfile(taxon_file):\n",
    "    with open(taxon_file, 'r') as f:\n",
    "        species_taxon_dict = json.load(f)\n",
    "        \n",
    "phylum_list = []\n",
    "subphylum_list = []\n",
    "class_list = []      \n",
    "for sp in db_sp_list:\n",
    "    phylum_list.append(species_taxon_dict[sp][\"Phylum\"])\n",
    "    subphylum_list.append(species_taxon_dict[sp][\"Subphylum\"])\n",
    "    class_list.append(species_taxon_dict[sp][\"Class\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color=#c994c7>MATEdb2: Martínez-Redondo et. al. 2024</font>\n",
    "For now I have no pipeline set up to extract sequences from the assembled transcriptomes but this may be coming in the future"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PHYLUM</th>\n",
       "      <th>LINEAGE</th>\n",
       "      <th>CLASS</th>\n",
       "      <th>ORDER</th>\n",
       "      <th>FAMILY</th>\n",
       "      <th>COMMON NAME</th>\n",
       "      <th>CODE (5 letter code)</th>\n",
       "      <th>SCIENTIFIC NAME</th>\n",
       "      <th>TAXON ID (see NCBI)</th>\n",
       "      <th>GENOME OR TRANSCRIPTOME?</th>\n",
       "      <th>...</th>\n",
       "      <th>C (LONGEST)</th>\n",
       "      <th>S (LONGEST)</th>\n",
       "      <th>D (LONGEST)</th>\n",
       "      <th>F (LONGEST)</th>\n",
       "      <th>M (LONGEST)</th>\n",
       "      <th>C+F (GOOD if &gt;85*).1</th>\n",
       "      <th>Busco database.1</th>\n",
       "      <th>No. GENES</th>\n",
       "      <th>EggNOG-mapper</th>\n",
       "      <th>FANTASIA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACOELA</td>\n",
       "      <td>Acoelomorpha</td>\n",
       "      <td>-</td>\n",
       "      <td>Acoela</td>\n",
       "      <td>Anaperidae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NFUS1</td>\n",
       "      <td>Neochildia fusca</td>\n",
       "      <td>66402</td>\n",
       "      <td>Transcriptome</td>\n",
       "      <td>...</td>\n",
       "      <td>74.7</td>\n",
       "      <td>72.7</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>22.3</td>\n",
       "      <td>77.7</td>\n",
       "      <td>metazoa_odb10</td>\n",
       "      <td>22370</td>\n",
       "      <td>12845</td>\n",
       "      <td>22351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACOELA</td>\n",
       "      <td>Acoelomorpha</td>\n",
       "      <td>-</td>\n",
       "      <td>Acoela</td>\n",
       "      <td>Childiidae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CSUB1</td>\n",
       "      <td>Childia submaculatum</td>\n",
       "      <td>243778</td>\n",
       "      <td>Transcriptome</td>\n",
       "      <td>...</td>\n",
       "      <td>71.7</td>\n",
       "      <td>65.9</td>\n",
       "      <td>5.8</td>\n",
       "      <td>5.3</td>\n",
       "      <td>23.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>metazoa_odb10</td>\n",
       "      <td>15861</td>\n",
       "      <td>10576</td>\n",
       "      <td>15605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACOELA</td>\n",
       "      <td>Acoelomorpha</td>\n",
       "      <td>-</td>\n",
       "      <td>Acoela</td>\n",
       "      <td>Convolutidae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CNAI1</td>\n",
       "      <td>Convoluta (Praesagittifera) naikaiensis</td>\n",
       "      <td>31270</td>\n",
       "      <td>Transcriptome</td>\n",
       "      <td>...</td>\n",
       "      <td>74.1</td>\n",
       "      <td>73.2</td>\n",
       "      <td>0.9</td>\n",
       "      <td>5.2</td>\n",
       "      <td>20.7</td>\n",
       "      <td>79.3</td>\n",
       "      <td>metazoa_odb10</td>\n",
       "      <td>17797</td>\n",
       "      <td>10900</td>\n",
       "      <td>17779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACOELA</td>\n",
       "      <td>Acoelomorpha</td>\n",
       "      <td>-</td>\n",
       "      <td>Acoela</td>\n",
       "      <td>Diopisthoporidae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DLON1</td>\n",
       "      <td>Diopisthoporus longitubus</td>\n",
       "      <td>996724</td>\n",
       "      <td>Transcriptome</td>\n",
       "      <td>...</td>\n",
       "      <td>71.6</td>\n",
       "      <td>64.8</td>\n",
       "      <td>6.8</td>\n",
       "      <td>3.7</td>\n",
       "      <td>24.7</td>\n",
       "      <td>75.3</td>\n",
       "      <td>metazoa_odb10</td>\n",
       "      <td>19356</td>\n",
       "      <td>13097</td>\n",
       "      <td>18921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACOELA</td>\n",
       "      <td>Acoelomorpha</td>\n",
       "      <td>-</td>\n",
       "      <td>Acoela</td>\n",
       "      <td>Hofsteniidae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HMIA1</td>\n",
       "      <td>Hofstenia miamia</td>\n",
       "      <td>442651</td>\n",
       "      <td>Genome</td>\n",
       "      <td>...</td>\n",
       "      <td>80.0</td>\n",
       "      <td>78.1</td>\n",
       "      <td>1.9</td>\n",
       "      <td>5.5</td>\n",
       "      <td>14.5</td>\n",
       "      <td>85.5</td>\n",
       "      <td>metazoa_odb10</td>\n",
       "      <td>22632</td>\n",
       "      <td>15809</td>\n",
       "      <td>21891</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   PHYLUM       LINEAGE CLASS    ORDER            FAMILY COMMON NAME  \\\n",
       "0  ACOELA  Acoelomorpha     -  Acoela         Anaperidae         NaN   \n",
       "1  ACOELA  Acoelomorpha     -  Acoela         Childiidae         NaN   \n",
       "2  ACOELA  Acoelomorpha     -  Acoela       Convolutidae         NaN   \n",
       "3  ACOELA  Acoelomorpha     -  Acoela   Diopisthoporidae         NaN   \n",
       "4  ACOELA  Acoelomorpha     -  Acoela       Hofsteniidae         NaN   \n",
       "\n",
       "  CODE (5 letter code)                          SCIENTIFIC NAME  \\\n",
       "0                NFUS1                         Neochildia fusca   \n",
       "1                CSUB1                     Childia submaculatum   \n",
       "2                CNAI1  Convoluta (Praesagittifera) naikaiensis   \n",
       "3                DLON1                Diopisthoporus longitubus   \n",
       "4                HMIA1                         Hofstenia miamia   \n",
       "\n",
       "   TAXON ID (see NCBI) GENOME OR TRANSCRIPTOME?  ... C (LONGEST) S (LONGEST)  \\\n",
       "0                66402            Transcriptome  ...        74.7        72.7   \n",
       "1               243778            Transcriptome  ...        71.7        65.9   \n",
       "2                31270            Transcriptome  ...        74.1        73.2   \n",
       "3               996724            Transcriptome  ...        71.6        64.8   \n",
       "4               442651                   Genome  ...        80.0        78.1   \n",
       "\n",
       "  D (LONGEST) F (LONGEST) M (LONGEST) C+F (GOOD if >85*).1 Busco database.1  \\\n",
       "0         2.0         3.0        22.3                 77.7    metazoa_odb10   \n",
       "1         5.8         5.3        23.0                 77.0    metazoa_odb10   \n",
       "2         0.9         5.2        20.7                 79.3    metazoa_odb10   \n",
       "3         6.8         3.7        24.7                 75.3    metazoa_odb10   \n",
       "4         1.9         5.5        14.5                 85.5    metazoa_odb10   \n",
       "\n",
       "  No. GENES EggNOG-mapper FANTASIA  \n",
       "0     22370         12845    22351  \n",
       "1     15861         10576    15605  \n",
       "2     17797         10900    17779  \n",
       "3     19356         13097    18921  \n",
       "4     22632         15809    21891  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matedb2_links_file = f'{report_dir2}/matedb2/Table_S1_cleaned.csv'\n",
    "matedb2_links_df = pd.read_csv(matedb2_links_file)\n",
    "matedb2_links_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(970, 33)"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matedb2_links_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PHYLUM</th>\n",
       "      <th>LINEAGE</th>\n",
       "      <th>CLASS</th>\n",
       "      <th>ORDER</th>\n",
       "      <th>FAMILY</th>\n",
       "      <th>COMMON NAME</th>\n",
       "      <th>CODE (5 letter code)</th>\n",
       "      <th>SCIENTIFIC NAME</th>\n",
       "      <th>TAXON ID (see NCBI)</th>\n",
       "      <th>GENOME OR TRANSCRIPTOME?</th>\n",
       "      <th>...</th>\n",
       "      <th>C (LONGEST)</th>\n",
       "      <th>S (LONGEST)</th>\n",
       "      <th>D (LONGEST)</th>\n",
       "      <th>F (LONGEST)</th>\n",
       "      <th>M (LONGEST)</th>\n",
       "      <th>C+F (GOOD if &gt;85*).1</th>\n",
       "      <th>Busco database.1</th>\n",
       "      <th>No. GENES</th>\n",
       "      <th>EggNOG-mapper</th>\n",
       "      <th>FANTASIA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ARTHROPODA</td>\n",
       "      <td>Chelicerata</td>\n",
       "      <td>??</td>\n",
       "      <td>Xiphosura</td>\n",
       "      <td>Limulidae</td>\n",
       "      <td>Horseshoe crabs</td>\n",
       "      <td>LPOL1</td>\n",
       "      <td>Limulus polyphemus</td>\n",
       "      <td>6850</td>\n",
       "      <td>Genome</td>\n",
       "      <td>...</td>\n",
       "      <td>92.5</td>\n",
       "      <td>86.0</td>\n",
       "      <td>6.5</td>\n",
       "      <td>4.7</td>\n",
       "      <td>2.8</td>\n",
       "      <td>97.2</td>\n",
       "      <td>metazoa_odb10</td>\n",
       "      <td>22879</td>\n",
       "      <td>20689</td>\n",
       "      <td>22686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ARTHROPODA</td>\n",
       "      <td>Chelicerata</td>\n",
       "      <td>Arachnida</td>\n",
       "      <td>Araneae</td>\n",
       "      <td>Araneidae</td>\n",
       "      <td>Spiders</td>\n",
       "      <td>ABRU1</td>\n",
       "      <td>Argiope bruennichi</td>\n",
       "      <td>94029</td>\n",
       "      <td>Genome</td>\n",
       "      <td>...</td>\n",
       "      <td>84.5</td>\n",
       "      <td>82.6</td>\n",
       "      <td>1.9</td>\n",
       "      <td>8.5</td>\n",
       "      <td>7.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>metazoa_odb10</td>\n",
       "      <td>23259</td>\n",
       "      <td>15836</td>\n",
       "      <td>23229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ARTHROPODA</td>\n",
       "      <td>Pancrustacea</td>\n",
       "      <td>Crustacea/Branchiopoda</td>\n",
       "      <td>Anostraca</td>\n",
       "      <td>Artemiidae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ASAL1</td>\n",
       "      <td>Artemia salina</td>\n",
       "      <td>85549</td>\n",
       "      <td>Transcriptome</td>\n",
       "      <td>...</td>\n",
       "      <td>63.2</td>\n",
       "      <td>63.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>15.7</td>\n",
       "      <td>21.1</td>\n",
       "      <td>78.9</td>\n",
       "      <td>metazoa_odb10</td>\n",
       "      <td>17880</td>\n",
       "      <td>12058</td>\n",
       "      <td>17869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ARTHROPODA</td>\n",
       "      <td>Pancrustacea</td>\n",
       "      <td>Crustacea/Branchiopoda</td>\n",
       "      <td>Diplostraca</td>\n",
       "      <td>Daphniidae</td>\n",
       "      <td>Water fleas</td>\n",
       "      <td>DMAG2</td>\n",
       "      <td>Daphnia magna</td>\n",
       "      <td>35525</td>\n",
       "      <td>Genome</td>\n",
       "      <td>...</td>\n",
       "      <td>96.8</td>\n",
       "      <td>90.6</td>\n",
       "      <td>6.2</td>\n",
       "      <td>0.3</td>\n",
       "      <td>2.9</td>\n",
       "      <td>97.1</td>\n",
       "      <td>metazoa_odb10</td>\n",
       "      <td>16891</td>\n",
       "      <td>14670</td>\n",
       "      <td>16206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ARTHROPODA</td>\n",
       "      <td>Pancrustacea</td>\n",
       "      <td>Crustacea/Malacostraca</td>\n",
       "      <td>Amphipoda</td>\n",
       "      <td>Talitridae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TASA1</td>\n",
       "      <td>Talitrus saltator</td>\n",
       "      <td>191375</td>\n",
       "      <td>Transcriptome</td>\n",
       "      <td>...</td>\n",
       "      <td>67.8</td>\n",
       "      <td>67.3</td>\n",
       "      <td>0.5</td>\n",
       "      <td>13.1</td>\n",
       "      <td>19.1</td>\n",
       "      <td>80.9</td>\n",
       "      <td>metazoa_odb10</td>\n",
       "      <td>15927</td>\n",
       "      <td>10006</td>\n",
       "      <td>15912</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       PHYLUM       LINEAGE                   CLASS        ORDER      FAMILY  \\\n",
       "0  ARTHROPODA   Chelicerata                      ??    Xiphosura   Limulidae   \n",
       "1  ARTHROPODA   Chelicerata               Arachnida      Araneae   Araneidae   \n",
       "2  ARTHROPODA  Pancrustacea  Crustacea/Branchiopoda    Anostraca  Artemiidae   \n",
       "3  ARTHROPODA  Pancrustacea  Crustacea/Branchiopoda  Diplostraca  Daphniidae   \n",
       "4  ARTHROPODA  Pancrustacea  Crustacea/Malacostraca    Amphipoda  Talitridae   \n",
       "\n",
       "       COMMON NAME CODE (5 letter code)     SCIENTIFIC NAME  \\\n",
       "0  Horseshoe crabs                LPOL1  Limulus polyphemus   \n",
       "1          Spiders                ABRU1  Argiope bruennichi   \n",
       "2              NaN                ASAL1      Artemia salina   \n",
       "3      Water fleas                DMAG2       Daphnia magna   \n",
       "4              NaN                TASA1   Talitrus saltator   \n",
       "\n",
       "   TAXON ID (see NCBI) GENOME OR TRANSCRIPTOME?  ... C (LONGEST) S (LONGEST)  \\\n",
       "0                 6850                   Genome  ...        92.5        86.0   \n",
       "1                94029                   Genome  ...        84.5        82.6   \n",
       "2                85549            Transcriptome  ...        63.2        63.0   \n",
       "3                35525                   Genome  ...        96.8        90.6   \n",
       "4               191375            Transcriptome  ...        67.8        67.3   \n",
       "\n",
       "  D (LONGEST) F (LONGEST) M (LONGEST) C+F (GOOD if >85*).1 Busco database.1  \\\n",
       "0         6.5         4.7         2.8                 97.2    metazoa_odb10   \n",
       "1         1.9         8.5         7.0                 93.0    metazoa_odb10   \n",
       "2         0.2        15.7        21.1                 78.9    metazoa_odb10   \n",
       "3         6.2         0.3         2.9                 97.1    metazoa_odb10   \n",
       "4         0.5        13.1        19.1                 80.9    metazoa_odb10   \n",
       "\n",
       "  No. GENES EggNOG-mapper FANTASIA  \n",
       "0     22879         20689    22686  \n",
       "1     23259         15836    23229  \n",
       "2     17880         12058    17869  \n",
       "3     16891         14670    16206  \n",
       "4     15927         10006    15912  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matedb2_links_filtered = matedb2_links_df[matedb2_links_df['SCIENTIFIC NAME'].isin(species_list)].reset_index(drop=True)\n",
    "matedb2_links_filtered.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(37, 33)"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matedb2_links_filtered.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(14, 33)"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matedb2_links_filtered2 = matedb2_links_filtered[~matedb2_links_filtered['SCIENTIFIC NAME'].isin(ncbi_sp_list)]\n",
    "matedb2_links_filtered2.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color=#c994c7>Create Merged Dataframe From All Accessory Opsin Sequence DBs</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Subphylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>Full_Species</th>\n",
       "      <th>Gene_Description</th>\n",
       "      <th>Protein</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>U3IU26</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Craniata</td>\n",
       "      <td>Aves</td>\n",
       "      <td>Anas</td>\n",
       "      <td>platyrhynchos</td>\n",
       "      <td>Anas platyrhynchos</td>\n",
       "      <td>Uncharacterized protein</td>\n",
       "      <td>TQTEHNIVAAYLITAGVISIFSNIVVLGIFVKYKELRTATNAIIIN...</td>\n",
       "      <td>Ramirez et al 2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0R4IEG0</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Craniata</td>\n",
       "      <td>Actinopteri</td>\n",
       "      <td>Danio</td>\n",
       "      <td>rerio</td>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>Novopsin-10</td>\n",
       "      <td>SPSADLSIAVFLIITGVVSVFGNGLVLLVYGQRRKKLRAHELMTIN...</td>\n",
       "      <td>Ramirez et al 2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q80XL3</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Craniata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mus</td>\n",
       "      <td>musculus</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Rrh protein</td>\n",
       "      <td>SRTEHSVIAAYLIVAGITSILSNVVVLGIFIKYKELRTPTNAVIIN...</td>\n",
       "      <td>Ramirez et al 2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M3YAY1</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Craniata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mustela</td>\n",
       "      <td>putorius furo</td>\n",
       "      <td>Mustela putorius furo</td>\n",
       "      <td>Uncharacterized protein</td>\n",
       "      <td>SQTEHNIVAAYLITAGIISIFSNLIVLGIFIKYKELRTPTNAIIIN...</td>\n",
       "      <td>Ramirez et al 2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G1PJY3</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Craniata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Myotis</td>\n",
       "      <td>lucifugus</td>\n",
       "      <td>Myotis lucifugus</td>\n",
       "      <td>Uncharacterized protein</td>\n",
       "      <td>SQTEHNIVATYLIMAGMISLLSNIIVLGIFITYKELRTPTNAIIIN...</td>\n",
       "      <td>Ramirez et al 2016</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Accession    Phylum Subphylum        Class    Genus        Species  \\\n",
       "0      U3IU26  Chordata  Craniata         Aves     Anas  platyrhynchos   \n",
       "1  A0A0R4IEG0  Chordata  Craniata  Actinopteri    Danio          rerio   \n",
       "2      Q80XL3  Chordata  Craniata     Mammalia      Mus       musculus   \n",
       "3      M3YAY1  Chordata  Craniata     Mammalia  Mustela  putorius furo   \n",
       "4      G1PJY3  Chordata  Craniata     Mammalia   Myotis      lucifugus   \n",
       "\n",
       "            Full_Species         Gene_Description  \\\n",
       "0     Anas platyrhynchos  Uncharacterized protein   \n",
       "1            Danio rerio              Novopsin-10   \n",
       "2           Mus musculus              Rrh protein   \n",
       "3  Mustela putorius furo  Uncharacterized protein   \n",
       "4       Myotis lucifugus  Uncharacterized protein   \n",
       "\n",
       "                                             Protein              source  \n",
       "0  TQTEHNIVAAYLITAGVISIFSNIVVLGIFVKYKELRTATNAIIIN...  Ramirez et al 2016  \n",
       "1  SPSADLSIAVFLIITGVVSVFGNGLVLLVYGQRRKKLRAHELMTIN...  Ramirez et al 2016  \n",
       "2  SRTEHSVIAAYLIVAGITSILSNVVVLGIFIKYKELRTPTNAVIIN...  Ramirez et al 2016  \n",
       "3  SQTEHNIVAAYLITAGIISIFSNLIVLGIFIKYKELRTPTNAIIIN...  Ramirez et al 2016  \n",
       "4  SQTEHNIVATYLIMAGMISLLSNIIVLGIFITYKELRTPTNAIIIN...  Ramirez et al 2016  "
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#make a merged df of all the accessory seq dbs, filter out reedundant datapoints, then append to the end of the NCBI query sheet?\n",
    "data = {'Accession': acc_list, 'Phylum': phylum_list, 'Subphylum': subphylum_list, 'Class': class_list,'Genus': gn_list, 'Species': sp_list, 'Full_Species': db_sp_list, 'Gene_Description': prot_descriptions, 'Protein' : sequence_list, 'source' : source_list} \n",
    "#data = {'Accession': acc_list, 'Genus': gn_list, 'Species': sp_list, 'Full_Species': db_sp_list, 'Gene_Description': prot_descriptions, 'Protein' : sequence_list, 'source' : source_list} \n",
    "\n",
    "acc_seq_db_df = pd.DataFrame(data)\n",
    "acc_seq_db_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2225, 10)"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "acc_seq_db_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2214, 10)"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "acc_seq_db_df_filtered = acc_seq_db_df.copy()\n",
    "acc_seq_db_df_filtered.drop_duplicates(subset=['Full_Species', 'Protein'],  keep='first', inplace=True)\n",
    "acc_seq_db_df_filtered=acc_seq_db_df_filtered.reset_index(drop=True)\n",
    "acc_seq_db_df_filtered.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "87"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(acc_seq_db_df_filtered['Full_Species'].to_list()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "acc_seq_db_df_filtered.to_csv(f'{report_dir2}/vpod_comp_accessory_seq_dbs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_file = f'./{report_dir2}/acc_db_seqs.fasta'\n",
    "with open(fasta_file, 'w') as f:\n",
    "    for id, seq in zip(acc_seq_db_df_filtered['Accession'], acc_seq_db_df_filtered['Protein']):\n",
    "        f.write(f'>{id}\\n{seq}\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color=#c994c7>Merge the Formated Accessory Sequence DBs w/the Mined NCBI Data</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Subphylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>Full_Species</th>\n",
       "      <th>DNA</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Gene_Description</th>\n",
       "      <th>Species_Synonym_Used</th>\n",
       "      <th>Prot_Len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XM_024085765.2</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexapoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Bicyclus</td>\n",
       "      <td>anynana</td>\n",
       "      <td>Bicyclus anynana</td>\n",
       "      <td>ACACAATTCATTTAACGTGTTCAGAATAGATGCACCACCGCATTAG...</td>\n",
       "      <td>MATQSCFWSDIFDEQSRANLNTGFVVTGAVSTLLGSWLNLALLLTN...</td>\n",
       "      <td>PREDICTED: Bicyclus anynana rhodopsin (LOC1120...</td>\n",
       "      <td>NA</td>\n",
       "      <td>313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XM_024089198.2</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexapoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Bicyclus</td>\n",
       "      <td>anynana</td>\n",
       "      <td>Bicyclus anynana</td>\n",
       "      <td>GTTCGGGATGACAGATATACGCTCGAGAATAACATTTTGCTACGGG...</td>\n",
       "      <td>MSPATKIIVIIFLIDIKCKNVDTLPEVKAELWSFKNNLKKSVVNHF...</td>\n",
       "      <td>PREDICTED: Bicyclus anynana opsin, ultraviolet...</td>\n",
       "      <td>NA</td>\n",
       "      <td>572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AY918895.2</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexapoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Bicyclus</td>\n",
       "      <td>anynana</td>\n",
       "      <td>Bicyclus anynana</td>\n",
       "      <td>TTTGGTATTAACCATCTCGGGGCGACCTATCCACCTGTAACCCTTT...</td>\n",
       "      <td>MAITSMDPGPGIAALQAWGGHAAAYGSNETVVDKVLPDMLHLVDPH...</td>\n",
       "      <td>Bicyclus anynana long wavelength sensitive ops...</td>\n",
       "      <td>NA</td>\n",
       "      <td>380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XM_061286956.1</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Craniata</td>\n",
       "      <td>Actinopteri</td>\n",
       "      <td>Syngnathus</td>\n",
       "      <td>typhle</td>\n",
       "      <td>Syngnathus typhle</td>\n",
       "      <td>ACCGCACGCGTGGATTACATTGCAGCGTGGAGGAAATGGGGAAAAA...</td>\n",
       "      <td>MDARMFSGQVALNLSGSNLSDDPELAEPSVEVEARLSPTGFVVLSV...</td>\n",
       "      <td>PREDICTED: Syngnathus typhle teleost multiple ...</td>\n",
       "      <td>NA</td>\n",
       "      <td>381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XM_061286026.1</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Craniata</td>\n",
       "      <td>Actinopteri</td>\n",
       "      <td>Syngnathus</td>\n",
       "      <td>typhle</td>\n",
       "      <td>Syngnathus typhle</td>\n",
       "      <td>GGACGCGCGCTCCAGCCGCACATCGAAGCCACCACAATGACCGAAA...</td>\n",
       "      <td>MNPYNETRAEHNLFAFGTYKLLAVTIGTIGVFGFCNNVLVILLYCK...</td>\n",
       "      <td>PREDICTED: Syngnathus typhle opsin 3 (opn3), mRNA</td>\n",
       "      <td>NA</td>\n",
       "      <td>381</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Accession      Phylum Subphylum        Class       Genus  Species  \\\n",
       "0  XM_024085765.2  Arthropoda  Hexapoda      Insecta    Bicyclus  anynana   \n",
       "1  XM_024089198.2  Arthropoda  Hexapoda      Insecta    Bicyclus  anynana   \n",
       "2      AY918895.2  Arthropoda  Hexapoda      Insecta    Bicyclus  anynana   \n",
       "3  XM_061286956.1    Chordata  Craniata  Actinopteri  Syngnathus   typhle   \n",
       "4  XM_061286026.1    Chordata  Craniata  Actinopteri  Syngnathus   typhle   \n",
       "\n",
       "        Full_Species                                                DNA  \\\n",
       "0   Bicyclus anynana  ACACAATTCATTTAACGTGTTCAGAATAGATGCACCACCGCATTAG...   \n",
       "1   Bicyclus anynana  GTTCGGGATGACAGATATACGCTCGAGAATAACATTTTGCTACGGG...   \n",
       "2   Bicyclus anynana  TTTGGTATTAACCATCTCGGGGCGACCTATCCACCTGTAACCCTTT...   \n",
       "3  Syngnathus typhle  ACCGCACGCGTGGATTACATTGCAGCGTGGAGGAAATGGGGAAAAA...   \n",
       "4  Syngnathus typhle  GGACGCGCGCTCCAGCCGCACATCGAAGCCACCACAATGACCGAAA...   \n",
       "\n",
       "                                             Protein  \\\n",
       "0  MATQSCFWSDIFDEQSRANLNTGFVVTGAVSTLLGSWLNLALLLTN...   \n",
       "1  MSPATKIIVIIFLIDIKCKNVDTLPEVKAELWSFKNNLKKSVVNHF...   \n",
       "2  MAITSMDPGPGIAALQAWGGHAAAYGSNETVVDKVLPDMLHLVDPH...   \n",
       "3  MDARMFSGQVALNLSGSNLSDDPELAEPSVEVEARLSPTGFVVLSV...   \n",
       "4  MNPYNETRAEHNLFAFGTYKLLAVTIGTIGVFGFCNNVLVILLYCK...   \n",
       "\n",
       "                                    Gene_Description Species_Synonym_Used  \\\n",
       "0  PREDICTED: Bicyclus anynana rhodopsin (LOC1120...                   NA   \n",
       "1  PREDICTED: Bicyclus anynana opsin, ultraviolet...                   NA   \n",
       "2  Bicyclus anynana long wavelength sensitive ops...                   NA   \n",
       "3  PREDICTED: Syngnathus typhle teleost multiple ...                   NA   \n",
       "4  PREDICTED: Syngnathus typhle opsin 3 (opn3), mRNA                   NA   \n",
       "\n",
       "   Prot_Len  \n",
       "0       313  \n",
       "1       572  \n",
       "2       380  \n",
       "3       381  \n",
       "4       381  "
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_query_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Subphylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>Full_Species</th>\n",
       "      <th>Gene_Description</th>\n",
       "      <th>Protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>U3IU26</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Craniata</td>\n",
       "      <td>Aves</td>\n",
       "      <td>Anas</td>\n",
       "      <td>platyrhynchos</td>\n",
       "      <td>Anas platyrhynchos</td>\n",
       "      <td>Uncharacterized protein</td>\n",
       "      <td>TQTEHNIVAAYLITAGVISIFSNIVVLGIFVKYKELRTATNAIIIN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0R4IEG0</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Craniata</td>\n",
       "      <td>Actinopteri</td>\n",
       "      <td>Danio</td>\n",
       "      <td>rerio</td>\n",
       "      <td>Danio rerio</td>\n",
       "      <td>Novopsin-10</td>\n",
       "      <td>SPSADLSIAVFLIITGVVSVFGNGLVLLVYGQRRKKLRAHELMTIN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q80XL3</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Craniata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mus</td>\n",
       "      <td>musculus</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Rrh protein</td>\n",
       "      <td>SRTEHSVIAAYLIVAGITSILSNVVVLGIFIKYKELRTPTNAVIIN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M3YAY1</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Craniata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Mustela</td>\n",
       "      <td>putorius furo</td>\n",
       "      <td>Mustela putorius furo</td>\n",
       "      <td>Uncharacterized protein</td>\n",
       "      <td>SQTEHNIVAAYLITAGIISIFSNLIVLGIFIKYKELRTPTNAIIIN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G1PJY3</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Craniata</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>Myotis</td>\n",
       "      <td>lucifugus</td>\n",
       "      <td>Myotis lucifugus</td>\n",
       "      <td>Uncharacterized protein</td>\n",
       "      <td>SQTEHNIVATYLIMAGMISLLSNIIVLGIFITYKELRTPTNAIIIN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Accession    Phylum Subphylum        Class    Genus        Species  \\\n",
       "0      U3IU26  Chordata  Craniata         Aves     Anas  platyrhynchos   \n",
       "1  A0A0R4IEG0  Chordata  Craniata  Actinopteri    Danio          rerio   \n",
       "2      Q80XL3  Chordata  Craniata     Mammalia      Mus       musculus   \n",
       "3      M3YAY1  Chordata  Craniata     Mammalia  Mustela  putorius furo   \n",
       "4      G1PJY3  Chordata  Craniata     Mammalia   Myotis      lucifugus   \n",
       "\n",
       "            Full_Species         Gene_Description  \\\n",
       "0     Anas platyrhynchos  Uncharacterized protein   \n",
       "1            Danio rerio              Novopsin-10   \n",
       "2           Mus musculus              Rrh protein   \n",
       "3  Mustela putorius furo  Uncharacterized protein   \n",
       "4       Myotis lucifugus  Uncharacterized protein   \n",
       "\n",
       "                                             Protein  \n",
       "0  TQTEHNIVAAYLITAGVISIFSNIVVLGIFVKYKELRTATNAIIIN...  \n",
       "1  SPSADLSIAVFLIITGVVSVFGNGLVLLVYGQRRKKLRAHELMTIN...  \n",
       "2  SRTEHSVIAAYLIVAGITSILSNVVVLGIFIKYKELRTPTNAVIIN...  \n",
       "3  SQTEHNIVAAYLITAGIISIFSNLIVLGIFIKYKELRTPTNAIIIN...  \n",
       "4  SQTEHNIVATYLIMAGMISLLSNIIVLGIFITYKELRTPTNAIIIN...  "
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "acc_seq_db_df_filtered.drop(columns='source', inplace =True)\n",
    "acc_seq_db_df_filtered.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Subphylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>Full_Species</th>\n",
       "      <th>DNA</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Gene_Description</th>\n",
       "      <th>Species_Synonym_Used</th>\n",
       "      <th>Prot_Len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XM_024085765.2</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexapoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Bicyclus</td>\n",
       "      <td>anynana</td>\n",
       "      <td>Bicyclus anynana</td>\n",
       "      <td>ACACAATTCATTTAACGTGTTCAGAATAGATGCACCACCGCATTAG...</td>\n",
       "      <td>MATQSCFWSDIFDEQSRANLNTGFVVTGAVSTLLGSWLNLALLLTN...</td>\n",
       "      <td>PREDICTED: Bicyclus anynana rhodopsin (LOC1120...</td>\n",
       "      <td>NA</td>\n",
       "      <td>313.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XM_024089198.2</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexapoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Bicyclus</td>\n",
       "      <td>anynana</td>\n",
       "      <td>Bicyclus anynana</td>\n",
       "      <td>GTTCGGGATGACAGATATACGCTCGAGAATAACATTTTGCTACGGG...</td>\n",
       "      <td>MSPATKIIVIIFLIDIKCKNVDTLPEVKAELWSFKNNLKKSVVNHF...</td>\n",
       "      <td>PREDICTED: Bicyclus anynana opsin, ultraviolet...</td>\n",
       "      <td>NA</td>\n",
       "      <td>572.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AY918895.2</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexapoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Bicyclus</td>\n",
       "      <td>anynana</td>\n",
       "      <td>Bicyclus anynana</td>\n",
       "      <td>TTTGGTATTAACCATCTCGGGGCGACCTATCCACCTGTAACCCTTT...</td>\n",
       "      <td>MAITSMDPGPGIAALQAWGGHAAAYGSNETVVDKVLPDMLHLVDPH...</td>\n",
       "      <td>Bicyclus anynana long wavelength sensitive ops...</td>\n",
       "      <td>NA</td>\n",
       "      <td>380.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XM_061286956.1</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Craniata</td>\n",
       "      <td>Actinopteri</td>\n",
       "      <td>Syngnathus</td>\n",
       "      <td>typhle</td>\n",
       "      <td>Syngnathus typhle</td>\n",
       "      <td>ACCGCACGCGTGGATTACATTGCAGCGTGGAGGAAATGGGGAAAAA...</td>\n",
       "      <td>MDARMFSGQVALNLSGSNLSDDPELAEPSVEVEARLSPTGFVVLSV...</td>\n",
       "      <td>PREDICTED: Syngnathus typhle teleost multiple ...</td>\n",
       "      <td>NA</td>\n",
       "      <td>381.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XM_061286026.1</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Craniata</td>\n",
       "      <td>Actinopteri</td>\n",
       "      <td>Syngnathus</td>\n",
       "      <td>typhle</td>\n",
       "      <td>Syngnathus typhle</td>\n",
       "      <td>GGACGCGCGCTCCAGCCGCACATCGAAGCCACCACAATGACCGAAA...</td>\n",
       "      <td>MNPYNETRAEHNLFAFGTYKLLAVTIGTIGVFGFCNNVLVILLYCK...</td>\n",
       "      <td>PREDICTED: Syngnathus typhle opsin 3 (opn3), mRNA</td>\n",
       "      <td>NA</td>\n",
       "      <td>381.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Accession      Phylum Subphylum        Class       Genus  Species  \\\n",
       "0  XM_024085765.2  Arthropoda  Hexapoda      Insecta    Bicyclus  anynana   \n",
       "1  XM_024089198.2  Arthropoda  Hexapoda      Insecta    Bicyclus  anynana   \n",
       "2      AY918895.2  Arthropoda  Hexapoda      Insecta    Bicyclus  anynana   \n",
       "3  XM_061286956.1    Chordata  Craniata  Actinopteri  Syngnathus   typhle   \n",
       "4  XM_061286026.1    Chordata  Craniata  Actinopteri  Syngnathus   typhle   \n",
       "\n",
       "        Full_Species                                                DNA  \\\n",
       "0   Bicyclus anynana  ACACAATTCATTTAACGTGTTCAGAATAGATGCACCACCGCATTAG...   \n",
       "1   Bicyclus anynana  GTTCGGGATGACAGATATACGCTCGAGAATAACATTTTGCTACGGG...   \n",
       "2   Bicyclus anynana  TTTGGTATTAACCATCTCGGGGCGACCTATCCACCTGTAACCCTTT...   \n",
       "3  Syngnathus typhle  ACCGCACGCGTGGATTACATTGCAGCGTGGAGGAAATGGGGAAAAA...   \n",
       "4  Syngnathus typhle  GGACGCGCGCTCCAGCCGCACATCGAAGCCACCACAATGACCGAAA...   \n",
       "\n",
       "                                             Protein  \\\n",
       "0  MATQSCFWSDIFDEQSRANLNTGFVVTGAVSTLLGSWLNLALLLTN...   \n",
       "1  MSPATKIIVIIFLIDIKCKNVDTLPEVKAELWSFKNNLKKSVVNHF...   \n",
       "2  MAITSMDPGPGIAALQAWGGHAAAYGSNETVVDKVLPDMLHLVDPH...   \n",
       "3  MDARMFSGQVALNLSGSNLSDDPELAEPSVEVEARLSPTGFVVLSV...   \n",
       "4  MNPYNETRAEHNLFAFGTYKLLAVTIGTIGVFGFCNNVLVILLYCK...   \n",
       "\n",
       "                                    Gene_Description Species_Synonym_Used  \\\n",
       "0  PREDICTED: Bicyclus anynana rhodopsin (LOC1120...                   NA   \n",
       "1  PREDICTED: Bicyclus anynana opsin, ultraviolet...                   NA   \n",
       "2  Bicyclus anynana long wavelength sensitive ops...                   NA   \n",
       "3  PREDICTED: Syngnathus typhle teleost multiple ...                   NA   \n",
       "4  PREDICTED: Syngnathus typhle opsin 3 (opn3), mRNA                   NA   \n",
       "\n",
       "   Prot_Len  \n",
       "0     313.0  \n",
       "1     572.0  \n",
       "2     380.0  \n",
       "3     381.0  \n",
       "4     381.0  "
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_query_df = pd.concat([ncbi_query_df, acc_seq_db_df_filtered]).reset_index(drop = True)\n",
    "final_query_df.to_csv(f'{query_report_dir}/ncbi_q_merged_w_acc_seq_db.csv', index=False)\n",
    "#final_query_df.to_csv('mnm_data/mnm_on_mnm_on_all_dbs_2024-12-10_23-31-03/ncbi_q_merged_w_acc_seq_db.csv')\n",
    "final_query_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4171, 12)"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_query_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_file = f'./{query_report_dir}/mined_and_acc_seqs.fasta'\n",
    "with open(fasta_file, 'w') as f:\n",
    "    for id, seq in zip(final_query_df['Accession'], final_query_df['Protein']):\n",
    "        f.write(f'>{id}\\n{seq}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'STOP' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[114], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[43mSTOP\u001b[49m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'STOP' is not defined"
     ]
    }
   ],
   "source": [
    "STOP"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make this section a notebook version of using OPTICS "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Optics function here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color=#c994c7>Part 2 Objective</font> - Match sequence to it's closest MSP value based on OPTICS predictions\n",
    "\n",
    "We'll need to...\n",
    "\n",
    "- Query OPTICS with all the sequence data [bootstrap enabled]\n",
    "- Extract Predictions\n",
    "- Match to closest MSP value species-by-species [will need a list of the unique species names] // Match to MaxId as the foreign key"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import time \n",
    "import datetime\n",
    "import warnings\n",
    "import pandas as pd\n",
    "from deepBreaks.preprocessing import read_data\n",
    "from mnm_scripts.mine_n_match_functions import mine_n_match\n",
    "email = 'sethfrazer@ucsb.edu'\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "warnings.simplefilter(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Subphylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>Full_Species</th>\n",
       "      <th>DNA</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Gene_Description</th>\n",
       "      <th>Species_Synonym_Used</th>\n",
       "      <th>Prot_Len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XM_024085765.2</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexapoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Bicyclus</td>\n",
       "      <td>anynana</td>\n",
       "      <td>Bicyclus anynana</td>\n",
       "      <td>ACACAATTCATTTAACGTGTTCAGAATAGATGCACCACCGCATTAG...</td>\n",
       "      <td>MATQSCFWSDIFDEQSRANLNTGFVVTGAVSTLLGSWLNLALLLTN...</td>\n",
       "      <td>PREDICTED: Bicyclus anynana rhodopsin (LOC1120...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>313.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XM_024089198.2</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexapoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Bicyclus</td>\n",
       "      <td>anynana</td>\n",
       "      <td>Bicyclus anynana</td>\n",
       "      <td>GTTCGGGATGACAGATATACGCTCGAGAATAACATTTTGCTACGGG...</td>\n",
       "      <td>MSPATKIIVIIFLIDIKCKNVDTLPEVKAELWSFKNNLKKSVVNHF...</td>\n",
       "      <td>PREDICTED: Bicyclus anynana opsin, ultraviolet...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>572.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AY918895.2</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Hexapoda</td>\n",
       "      <td>Insecta</td>\n",
       "      <td>Bicyclus</td>\n",
       "      <td>anynana</td>\n",
       "      <td>Bicyclus anynana</td>\n",
       "      <td>TTTGGTATTAACCATCTCGGGGCGACCTATCCACCTGTAACCCTTT...</td>\n",
       "      <td>MAITSMDPGPGIAALQAWGGHAAAYGSNETVVDKVLPDMLHLVDPH...</td>\n",
       "      <td>Bicyclus anynana long wavelength sensitive ops...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>380.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XM_061286956.1</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Craniata</td>\n",
       "      <td>Actinopteri</td>\n",
       "      <td>Syngnathus</td>\n",
       "      <td>typhle</td>\n",
       "      <td>Syngnathus typhle</td>\n",
       "      <td>ACCGCACGCGTGGATTACATTGCAGCGTGGAGGAAATGGGGAAAAA...</td>\n",
       "      <td>MDARMFSGQVALNLSGSNLSDDPELAEPSVEVEARLSPTGFVVLSV...</td>\n",
       "      <td>PREDICTED: Syngnathus typhle teleost multiple ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>381.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XM_061286026.1</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Craniata</td>\n",
       "      <td>Actinopteri</td>\n",
       "      <td>Syngnathus</td>\n",
       "      <td>typhle</td>\n",
       "      <td>Syngnathus typhle</td>\n",
       "      <td>GGACGCGCGCTCCAGCCGCACATCGAAGCCACCACAATGACCGAAA...</td>\n",
       "      <td>MNPYNETRAEHNLFAFGTYKLLAVTIGTIGVFGFCNNVLVILLYCK...</td>\n",
       "      <td>PREDICTED: Syngnathus typhle opsin 3 (opn3), mRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>381.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Accession      Phylum Subphylum        Class       Genus  Species  \\\n",
       "0  XM_024085765.2  Arthropoda  Hexapoda      Insecta    Bicyclus  anynana   \n",
       "1  XM_024089198.2  Arthropoda  Hexapoda      Insecta    Bicyclus  anynana   \n",
       "2      AY918895.2  Arthropoda  Hexapoda      Insecta    Bicyclus  anynana   \n",
       "3  XM_061286956.1    Chordata  Craniata  Actinopteri  Syngnathus   typhle   \n",
       "4  XM_061286026.1    Chordata  Craniata  Actinopteri  Syngnathus   typhle   \n",
       "\n",
       "        Full_Species                                                DNA  \\\n",
       "0   Bicyclus anynana  ACACAATTCATTTAACGTGTTCAGAATAGATGCACCACCGCATTAG...   \n",
       "1   Bicyclus anynana  GTTCGGGATGACAGATATACGCTCGAGAATAACATTTTGCTACGGG...   \n",
       "2   Bicyclus anynana  TTTGGTATTAACCATCTCGGGGCGACCTATCCACCTGTAACCCTTT...   \n",
       "3  Syngnathus typhle  ACCGCACGCGTGGATTACATTGCAGCGTGGAGGAAATGGGGAAAAA...   \n",
       "4  Syngnathus typhle  GGACGCGCGCTCCAGCCGCACATCGAAGCCACCACAATGACCGAAA...   \n",
       "\n",
       "                                             Protein  \\\n",
       "0  MATQSCFWSDIFDEQSRANLNTGFVVTGAVSTLLGSWLNLALLLTN...   \n",
       "1  MSPATKIIVIIFLIDIKCKNVDTLPEVKAELWSFKNNLKKSVVNHF...   \n",
       "2  MAITSMDPGPGIAALQAWGGHAAAYGSNETVVDKVLPDMLHLVDPH...   \n",
       "3  MDARMFSGQVALNLSGSNLSDDPELAEPSVEVEARLSPTGFVVLSV...   \n",
       "4  MNPYNETRAEHNLFAFGTYKLLAVTIGTIGVFGFCNNVLVILLYCK...   \n",
       "\n",
       "                                    Gene_Description Species_Synonym_Used  \\\n",
       "0  PREDICTED: Bicyclus anynana rhodopsin (LOC1120...                  NaN   \n",
       "1  PREDICTED: Bicyclus anynana opsin, ultraviolet...                  NaN   \n",
       "2  Bicyclus anynana long wavelength sensitive ops...                  NaN   \n",
       "3  PREDICTED: Syngnathus typhle teleost multiple ...                  NaN   \n",
       "4  PREDICTED: Syngnathus typhle opsin 3 (opn3), mRNA                  NaN   \n",
       "\n",
       "   Prot_Len  \n",
       "0     313.0  \n",
       "1     572.0  \n",
       "2     380.0  \n",
       "3     381.0  \n",
       "4     381.0  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "report_dir = 'mnm_data/mnm_on_mnm_on_all_dbs_2025-02-05_23-04-27' #re-define the report directory if needed\n",
    "ncbi_q_file = f'./{report_dir}/ncbi_q_merged_w_acc_seq_db.csv'\n",
    "ncbi = pd.read_csv(ncbi_q_file)\n",
    "ncbi.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4171, 12)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Names</th>\n",
       "      <th>Single_Prediction</th>\n",
       "      <th>Prediction_Means</th>\n",
       "      <th>Prediction_Medians</th>\n",
       "      <th>Prediction_Lower_Bounds</th>\n",
       "      <th>Prediction_Upper_Bounds</th>\n",
       "      <th>Std_Deviation</th>\n",
       "      <th>%Identity_Nearest_VPOD_Sequence</th>\n",
       "      <th>Sequence_Length</th>\n",
       "      <th>Lmax_Hex_Color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XM_024085765.2</td>\n",
       "      <td>436.6</td>\n",
       "      <td>439.5</td>\n",
       "      <td>440.9</td>\n",
       "      <td>396.4</td>\n",
       "      <td>477.4</td>\n",
       "      <td>22.8</td>\n",
       "      <td>29.814</td>\n",
       "      <td>313</td>\n",
       "      <td>#0600fe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XM_024089198.2</td>\n",
       "      <td>388.9</td>\n",
       "      <td>425.6</td>\n",
       "      <td>427.3</td>\n",
       "      <td>369.3</td>\n",
       "      <td>475.4</td>\n",
       "      <td>28.3</td>\n",
       "      <td>47.619</td>\n",
       "      <td>572</td>\n",
       "      <td>#4600dc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AY918895.2</td>\n",
       "      <td>549.2</td>\n",
       "      <td>538.2</td>\n",
       "      <td>538.1</td>\n",
       "      <td>518.0</td>\n",
       "      <td>554.4</td>\n",
       "      <td>10.1</td>\n",
       "      <td>81.746</td>\n",
       "      <td>380</td>\n",
       "      <td>#7bff00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XM_061286956.1</td>\n",
       "      <td>483.8</td>\n",
       "      <td>477.8</td>\n",
       "      <td>477.2</td>\n",
       "      <td>434.9</td>\n",
       "      <td>522.0</td>\n",
       "      <td>24.6</td>\n",
       "      <td>40.161</td>\n",
       "      <td>381</td>\n",
       "      <td>#00ccff</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XM_061286026.1</td>\n",
       "      <td>461.7</td>\n",
       "      <td>489.1</td>\n",
       "      <td>488.0</td>\n",
       "      <td>449.0</td>\n",
       "      <td>526.6</td>\n",
       "      <td>20.7</td>\n",
       "      <td>39.245</td>\n",
       "      <td>381</td>\n",
       "      <td>#00fbff</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Names  Single_Prediction  Prediction_Means  Prediction_Medians  \\\n",
       "0  XM_024085765.2              436.6             439.5               440.9   \n",
       "1  XM_024089198.2              388.9             425.6               427.3   \n",
       "2      AY918895.2              549.2             538.2               538.1   \n",
       "3  XM_061286956.1              483.8             477.8               477.2   \n",
       "4  XM_061286026.1              461.7             489.1               488.0   \n",
       "\n",
       "   Prediction_Lower_Bounds  Prediction_Upper_Bounds  Std_Deviation  \\\n",
       "0                    396.4                    477.4           22.8   \n",
       "1                    369.3                    475.4           28.3   \n",
       "2                    518.0                    554.4           10.1   \n",
       "3                    434.9                    522.0           24.6   \n",
       "4                    449.0                    526.6           20.7   \n",
       "\n",
       "  %Identity_Nearest_VPOD_Sequence  Sequence_Length Lmax_Hex_Color  \n",
       "0                          29.814              313        #0600fe  \n",
       "1                          47.619              572        #4600dc  \n",
       "2                          81.746              380        #7bff00  \n",
       "3                          40.161              381        #00ccff  \n",
       "4                          39.245              381        #00fbff  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Example use for OPTICS\n",
    "# python optics_predictions.py -in msp_mined_seqs.fasta -rd mined_msp_seqs -out mined_seq_predictions.tsv -m wildtype -e aa_prop -b True -ir msp_mined_seq_blastp_report.tsv -r bovine -s False -bsv msp_bs_viz.pdf\n",
    "pred_dir = 'e:\\safra\\Documents\\GitHub\\optics\\prediction_outputs\\optics_on_mnm_all_dbs_2025-02-06_23-25-42'\n",
    "optics_pred_file = f'{pred_dir}/mnm_all_dbs_predictions.tsv'\n",
    "optics = pd.read_csv(optics_pred_file, sep='\\t')\n",
    "\n",
    "optics.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4171, 10)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "optics.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Full_Species</th>\n",
       "      <th>Accession</th>\n",
       "      <th>maxid</th>\n",
       "      <th>longcore_id</th>\n",
       "      <th>murwes_id</th>\n",
       "      <th>caves_id</th>\n",
       "      <th>porter2005_id</th>\n",
       "      <th>porter2006_id</th>\n",
       "      <th>LambdaMax</th>\n",
       "      <th>Column1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>comp_db_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>Z86118</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42.0</td>\n",
       "      <td>508.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Loligo subulata</td>\n",
       "      <td>Z49108</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>499.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Schistocerca gregaria</td>\n",
       "      <td>X80072</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>46.0</td>\n",
       "      <td>430.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Schistocerca gregaria</td>\n",
       "      <td>X80071</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>39.0</td>\n",
       "      <td>520.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sphodromantis sp.</td>\n",
       "      <td>X71665</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>40.0</td>\n",
       "      <td>515.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Full_Species Accession  maxid  longcore_id  murwes_id  \\\n",
       "comp_db_id                                                                     \n",
       "0           Drosophila melanogaster    Z86118    NaN          NaN        NaN   \n",
       "1                   Loligo subulata    Z49108    NaN          NaN        NaN   \n",
       "2             Schistocerca gregaria    X80072    NaN          NaN        NaN   \n",
       "3             Schistocerca gregaria    X80071    NaN          NaN        NaN   \n",
       "4                 Sphodromantis sp.    X71665    NaN          NaN        NaN   \n",
       "\n",
       "            caves_id  porter2005_id  porter2006_id  LambdaMax  Column1  \n",
       "comp_db_id                                                              \n",
       "0                NaN            NaN           42.0      508.0      NaN  \n",
       "1                NaN            NaN            2.0      499.0      NaN  \n",
       "2                NaN            NaN           46.0      430.0      NaN  \n",
       "3                NaN            NaN           39.0      520.0      NaN  \n",
       "4                NaN            NaN           40.0      515.0      NaN  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "source_file = './data_sources/lmax/clean_vpod_comp_acc_dbs_2025-01-24_18-44-29_2.tsv'\n",
    "comp_db = pd.read_csv(source_file,delimiter='\\t',index_col=0)\n",
    "comp_db.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There were 0 unmatched species\n",
      "Existing Taxon Dictionary Found! One Moment While We Update It...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "final_err_filtered_df = mine_n_match(email, report_dir, source_file, ncbi_q_file, optics_pred_file, out='vpod_acc_dbs', err_filter = 15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Subphylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>Full_Species</th>\n",
       "      <th>%Identity_Nearest_VPOD_Sequence</th>\n",
       "      <th>prediction_value</th>\n",
       "      <th>LambdaMax</th>\n",
       "      <th>abs_diff</th>\n",
       "      <th>comp_db_id</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Gene_Description</th>\n",
       "      <th>Notes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mnm_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OU342715.1-13552739-13554338</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Craniata</td>\n",
       "      <td>Actinopteri</td>\n",
       "      <td>Taurulus</td>\n",
       "      <td>bubalis</td>\n",
       "      <td>Taurulus bubalis</td>\n",
       "      <td>87.755</td>\n",
       "      <td>510.7</td>\n",
       "      <td>507.00</td>\n",
       "      <td>3.70</td>\n",
       "      <td>1605</td>\n",
       "      <td>ANGTEGKNFYIPMSNRTGIVRSPFEYQQYYLADPIMFKLLAFYMFF...</td>\n",
       "      <td>rh2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STO|Pscil|UVS|contig03244</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Malacostraca</td>\n",
       "      <td>Pseudosquilla</td>\n",
       "      <td>ciliata</td>\n",
       "      <td>Pseudosquilla ciliata</td>\n",
       "      <td>55.172</td>\n",
       "      <td>409.6</td>\n",
       "      <td>403.17</td>\n",
       "      <td>6.43</td>\n",
       "      <td>646</td>\n",
       "      <td>APLPKLDALVALPPALLANLTLRIGRGIRHCSGPKGSVFASRYEVK...</td>\n",
       "      <td>SWS/UVS</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DEC|Nenor|LWS|TRINITY_DN39887_c5_g3_i2</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Malacostraca</td>\n",
       "      <td>Nephrops</td>\n",
       "      <td>norvegicus</td>\n",
       "      <td>Nephrops norvegicus</td>\n",
       "      <td>70.642</td>\n",
       "      <td>510.8</td>\n",
       "      <td>515.00</td>\n",
       "      <td>4.20</td>\n",
       "      <td>891</td>\n",
       "      <td>ASNNLPSTNPYGNYTVVDTVPKEILHMVDPHWYQFPPMNPLWYGLV...</td>\n",
       "      <td>LWS</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CM010448.1-19773550-19776969</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>Craniata</td>\n",
       "      <td>Actinopteri</td>\n",
       "      <td>Carassius</td>\n",
       "      <td>auratus</td>\n",
       "      <td>Carassius auratus</td>\n",
       "      <td>34.328</td>\n",
       "      <td>444.3</td>\n",
       "      <td>452.00</td>\n",
       "      <td>7.70</td>\n",
       "      <td>835</td>\n",
       "      <td>DIHVLNVTVYRVSNGGETAIGVYLVILGLSWIGNGVVILLLTKQRK...</td>\n",
       "      <td>opn6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DEC|Sydeb|MWS|KP234272.1</td>\n",
       "      <td>Arthropoda</td>\n",
       "      <td>Crustacea</td>\n",
       "      <td>Malacostraca</td>\n",
       "      <td>Systellaspis</td>\n",
       "      <td>debilis</td>\n",
       "      <td>Systellaspis debilis</td>\n",
       "      <td>40.659</td>\n",
       "      <td>483.8</td>\n",
       "      <td>493.00</td>\n",
       "      <td>9.20</td>\n",
       "      <td>765</td>\n",
       "      <td>DKMAGTQNASYAMSFTRGGSLISYGYPEGVTLMDLLPEDVKPLIHS...</td>\n",
       "      <td>MWS2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     Accession      Phylum  Subphylum  \\\n",
       "mnm_id                                                                  \n",
       "0                 OU342715.1-13552739-13554338    Chordata   Craniata   \n",
       "1                    STO|Pscil|UVS|contig03244  Arthropoda  Crustacea   \n",
       "2       DEC|Nenor|LWS|TRINITY_DN39887_c5_g3_i2  Arthropoda  Crustacea   \n",
       "3                 CM010448.1-19773550-19776969    Chordata   Craniata   \n",
       "4                     DEC|Sydeb|MWS|KP234272.1  Arthropoda  Crustacea   \n",
       "\n",
       "               Class          Genus     Species           Full_Species  \\\n",
       "mnm_id                                                                   \n",
       "0        Actinopteri       Taurulus     bubalis       Taurulus bubalis   \n",
       "1       Malacostraca  Pseudosquilla     ciliata  Pseudosquilla ciliata   \n",
       "2       Malacostraca       Nephrops  norvegicus    Nephrops norvegicus   \n",
       "3        Actinopteri      Carassius     auratus      Carassius auratus   \n",
       "4       Malacostraca   Systellaspis     debilis   Systellaspis debilis   \n",
       "\n",
       "       %Identity_Nearest_VPOD_Sequence  prediction_value  LambdaMax  abs_diff  \\\n",
       "mnm_id                                                                          \n",
       "0                               87.755             510.7     507.00      3.70   \n",
       "1                               55.172             409.6     403.17      6.43   \n",
       "2                               70.642             510.8     515.00      4.20   \n",
       "3                               34.328             444.3     452.00      7.70   \n",
       "4                               40.659             483.8     493.00      9.20   \n",
       "\n",
       "        comp_db_id                                            Protein  \\\n",
       "mnm_id                                                                  \n",
       "0             1605  ANGTEGKNFYIPMSNRTGIVRSPFEYQQYYLADPIMFKLLAFYMFF...   \n",
       "1              646  APLPKLDALVALPPALLANLTLRIGRGIRHCSGPKGSVFASRYEVK...   \n",
       "2              891  ASNNLPSTNPYGNYTVVDTVPKEILHMVDPHWYQFPPMNPLWYGLV...   \n",
       "3              835  DIHVLNVTVYRVSNGGETAIGVYLVILGLSWIGNGVVILLLTKQRK...   \n",
       "4              765  DKMAGTQNASYAMSFTRGGSLISYGYPEGVTLMDLLPEDVKPLIHS...   \n",
       "\n",
       "       Gene_Description Notes  \n",
       "mnm_id                         \n",
       "0                   rh2   NaN  \n",
       "1               SWS/UVS   NaN  \n",
       "2                   LWS   NaN  \n",
       "3                  opn6   NaN  \n",
       "4                  MWS2   NaN  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_err_filtered_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(498, 15)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_err_filtered_df.shape"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "deepBreaks",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.17"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
